{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "01de3647-f731-4da0-a12b-ea887e58cb16",
   "metadata": {},
   "source": [
    "##### imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1761,
   "id": "4110ccaa-793e-43f0-9d31-00afd6b76c06",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import requests\n",
    "import time\n",
    "import pprint\n",
    "from urllib.parse import quote_plus\n",
    "import re\n",
    "import json\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2c12bcf-8cce-4398-be55-bcc103525efb",
   "metadata": {},
   "source": [
    "### DFCI_2014_PES study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1762,
   "id": "0e3b6784-eadb-4f80-bf38-d376e31f0b4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get data\n",
    "#TODO import as objects from harvester\n",
    "## add parameter to skip first 4 lines in patient and study sample data\n",
    "\n",
    "init_mut_df = pd.read_csv('/Users/costellh/repos/metakb/hmc_notebooks/es_dfarber_broad_2014_extracted/es_dfarber_broad_2014/data_mutations.txt', sep='\\t')\n",
    "init_study_meta = pd.read_csv('/Users/costellh/repos/metakb/hmc_notebooks/es_dfarber_broad_2014_extracted/es_dfarber_broad_2014/meta_study.txt', sep='\\t')\n",
    "init_patient_df = pd.read_csv('/Users/costellh/repos/metakb/hmc_notebooks/es_dfarber_broad_2014_extracted/es_dfarber_broad_2014/data_clinical_patient.txt', sep='\\t', skiprows=4)\n",
    "init_sample_df = pd.read_csv('/Users/costellh/repos/metakb/hmc_notebooks/es_dfarber_broad_2014_extracted/es_dfarber_broad_2014/data_clinical_sample.txt', sep='\\t', skiprows=4)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ba8aeda-91a2-4cf0-9558-cf9292788936",
   "metadata": {},
   "source": [
    "#### Variant data: subset columns, check for duplicates "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1763,
   "id": "b19a85c8-3894-467e-9c4f-c38549a96f32",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of duplicate rows : 495\n",
      "\n",
      "Duplicate rows (excluding first instance):\n",
      "      Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
      "720         ABCA1          9       107607765     107607765   \n",
      "730         ABCA3         16         2338066       2338066   \n",
      "783         ABCC9         12        21954066      21954066   \n",
      "813         ACACA         17        35640173      35640173   \n",
      "829          ACHE          7       100490251     100490251   \n",
      "...           ...        ...             ...           ...   \n",
      "15134       STAG2          X       123179197     123179197   \n",
      "15155     ZDHHC15          X        74742823      74742824   \n",
      "15191       HUWE1          X        53579734      53579734   \n",
      "15216     SHROOM4          X        50350700      50350700   \n",
      "15218        SOX3          X       139586714     139586714   \n",
      "\n",
      "              Consequence Variant_Classification Variant_Type  \\\n",
      "720      missense_variant      Missense_Mutation          SNP   \n",
      "730      missense_variant      Missense_Mutation          SNP   \n",
      "783      missense_variant      Missense_Mutation          SNP   \n",
      "813      missense_variant      Missense_Mutation          SNP   \n",
      "829    synonymous_variant                 Silent          SNP   \n",
      "...                   ...                    ...          ...   \n",
      "15134         stop_gained      Nonsense_Mutation          SNP   \n",
      "15155  frameshift_variant        Frame_Shift_Ins          INS   \n",
      "15191    missense_variant      Missense_Mutation          SNP   \n",
      "15216    missense_variant      Missense_Mutation          SNP   \n",
      "15218    missense_variant      Missense_Mutation          SNP   \n",
      "\n",
      "      Reference_Allele Tumor_Seq_Allele2   SAMPLE_ID Sequence_Source  \\\n",
      "720                  G                 A  SJDES003-R             WXS   \n",
      "730                  C                 T  SJDES011-R             WXS   \n",
      "783                  C                 T    SJDES017             WXS   \n",
      "813                  T                 C    SJDES005             WXS   \n",
      "829                  G                 A     03-P152             WXS   \n",
      "...                ...               ...         ...             ...   \n",
      "15134                C                 T    SJDES006             WXS   \n",
      "15155                -                 C    SJDES010             WXS   \n",
      "15191                C                 A      TTC466             WXS   \n",
      "15216                C                 T      TTC466             WXS   \n",
      "15218                T                 C      TTC466             WXS   \n",
      "\n",
      "                             HGVSc              HGVSp  HGVSp_Short  \\\n",
      "720     ENST00000374736.3:c.806C>T        p.Ala269Val      p.A269V   \n",
      "730    ENST00000301732.5:c.2965G>A        p.Ala989Thr      p.A989T   \n",
      "783    ENST00000261200.4:c.4562G>A       p.Arg1521Gln     p.R1521Q   \n",
      "813     ENST00000353139.5:c.605A>G        p.Asn202Ser      p.N202S   \n",
      "829    ENST00000302913.4:c.1257C>T          p.Pro419=      p.P419=   \n",
      "...                            ...                ...          ...   \n",
      "15134   ENST00000218089.9:c.646C>T        p.Arg216Ter      p.R216*   \n",
      "15155    ENST00000373367.3:c.36dup  p.Leu13AlafsTer30  p.L13Afs*30   \n",
      "15191  ENST00000342160.3:c.8615G>T       p.Gly2872Val     p.G2872V   \n",
      "15216  ENST00000376020.2:c.3442G>A       p.Glu1148Lys     p.E1148K   \n",
      "15218   ENST00000370536.2:c.512A>G        p.Lys171Arg      p.K171R   \n",
      "\n",
      "         Transcript_ID          RefSeq  Protein_position  \n",
      "720    ENST00000374736     NM_005502.3             269.0  \n",
      "730    ENST00000301732     NM_001089.2             989.0  \n",
      "783    ENST00000261200     NM_020297.2            1521.0  \n",
      "813    ENST00000353139     NM_198834.1             202.0  \n",
      "829    ENST00000302913     NM_015831.2             419.0  \n",
      "...                ...             ...               ...  \n",
      "15134  ENST00000218089  NM_001042749.1             216.0  \n",
      "15155  ENST00000373367     NM_144969.2              12.0  \n",
      "15191  ENST00000342160             NaN            2872.0  \n",
      "15216  ENST00000376020     NM_020717.3            1148.0  \n",
      "15218  ENST00000370536     NM_005634.2             171.0  \n",
      "\n",
      "[495 rows x 17 columns]\n",
      "\n",
      "DataFrame shape after removing duplicates: (14737, 17)\n"
     ]
    }
   ],
   "source": [
    "# clean variant data\n",
    "\n",
    "# subset for necessary columns\n",
    "mut_df = init_mut_df.filter(['Hugo_Symbol',\n",
    "                        'Chromosome',\n",
    "                        'Start_Position', \n",
    "                        'End_Position',\n",
    "                        'Consequence',\n",
    "                        'Variant_Classification',\n",
    "                        'Variant_Type',\n",
    "                        'Reference_Allele',\n",
    "                        'Tumor_Seq_Allele2',\n",
    "                        'Tumor_Sample_Barcode',\n",
    "                        'Sequence_Source',\n",
    "                        'HGVSc',\n",
    "                        'HGVSp',\n",
    "                        'HGVSp_Short',\n",
    "                        'Transcript_ID',\n",
    "                        'RefSeq',\n",
    "                        'Protein_position'\n",
    "                       ], axis=1)\n",
    "\n",
    "\n",
    "# Strip whitespace and retry\n",
    "mut_df.columns = mut_df.columns.str.strip()\n",
    "mut_df = mut_df.rename(columns={'Tumor_Sample_Barcode': 'SAMPLE_ID'})\n",
    "\n",
    "# Check duplicate count\n",
    "num_duplicates = mut_df.duplicated().sum()\n",
    "print(f\"Number of duplicate rows : {num_duplicates}\")\n",
    "\n",
    "# TODO check duplicates for sanity check\n",
    "# print duplicates (excluding first instance)\n",
    "if num_duplicates > 0:\n",
    "    print(\"\\nDuplicate rows (excluding first instance):\")\n",
    "    print(mut_df[mut_df.duplicated()])\n",
    "\n",
    "# # print full duplicate groups (including the first occurrences)\n",
    "#     print(\"\\nAll rows involved in duplication:\")\n",
    "#     print(mut_df[mut_df.duplicated(keep=False)])\n",
    "\n",
    "# save duplicate rows to file\n",
    "    dupes = mut_df[mut_df.duplicated(keep=False)]\n",
    "    mut_df.to_csv('mut_dupes.csv', index=False)\n",
    "    \n",
    "\n",
    "    \n",
    "# remove duplicates, but keep first occurrence\n",
    "    mut_df = mut_df.drop_duplicates()\n",
    "    print(f\"\\nDataFrame shape after removing duplicates: {mut_df.shape}\")\n",
    "else:\n",
    "    print(\"No duplicate rows found.\")\n",
    "\n",
    "    \n",
    "# print(mut_df.columns)\n",
    "# # print(mut_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "780f5ee3-c672-4e26-b670-576086526f13",
   "metadata": {},
   "source": [
    "#### Patient data: subset columns, check for duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1764,
   "id": "75f3d6c8-9257-4ac6-b883-75a967ba8592",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['PATIENT_ID', 'AGE', 'SEX', 'ETHNICITY'], dtype='object')\n",
      "Number of duplicate rows : 0\n",
      "No duplicate rows found.\n"
     ]
    }
   ],
   "source": [
    "# clean patient data\n",
    "# print(patient_df.shape)\n",
    "# print(patient_df.columns)\n",
    "\n",
    "# subset data\n",
    "# subset for necessary columns\n",
    "patient_df = init_patient_df.filter(['PATIENT_ID',\n",
    "                        'AGE',\n",
    "                        'SEX', \n",
    "                        'ETHNICITY',\n",
    "                        'Consequence'\n",
    "                       ], axis=1)\n",
    "\n",
    "print(patient_df.columns)\n",
    "\n",
    "# Check duplicate count\n",
    "num_duplicates = patient_df.duplicated().sum()\n",
    "print(f\"Number of duplicate rows : {num_duplicates}\")\n",
    "\n",
    "# print duplicates (excluding first instance)\n",
    "if num_duplicates > 0:\n",
    "    print(\"\\nDuplicate rows (excluding first instance):\")\n",
    "    print(patient_df[patient_df.duplicated()])\n",
    "\n",
    "# # print full duplicate groups (including the first occurrences)\n",
    "#     print(\"\\nAll rows involved in duplication:\")\n",
    "#     print(mut_df[mut_df.duplicated(keep=False)])\n",
    "\n",
    "# remove duplicates, but keep first occurrence\n",
    "    patient_df = patient_df.drop_duplicates()\n",
    "    print(f\"\\nDataFrame shape after removing duplicates: {patient_df.shape}\")\n",
    "else:\n",
    "    print(\"No duplicate rows found.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0da33e63-f38d-4d9e-abdc-e3c0bba84675",
   "metadata": {},
   "source": [
    "#### Sample data: subset columns, check for duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1765,
   "id": "fb2499e2-affb-404e-b970-04aa20bef06e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['PATIENT_ID', 'SAMPLE_ID', 'SAMPLE_CLASS', 'ONCOTREE_CODE',\n",
      "       'CANCER_TYPE', 'CANCER_TYPE_DETAILED', 'TMB_NONSYNONYMOUS'],\n",
      "      dtype='object')\n",
      "Number of duplicate rows : 0\n",
      "No duplicate rows found.\n"
     ]
    }
   ],
   "source": [
    "# clean sample data\n",
    "\n",
    "# print(sample_df.shape)\n",
    "# print(sample_df.columns)\n",
    "# print(sample_df.head())\n",
    "\n",
    "\n",
    "# subset data\n",
    "# subset for necessary columns\n",
    "sample_df = init_sample_df.filter(['PATIENT_ID',\n",
    "                                     'SAMPLE_ID',\n",
    "                                     'SAMPLE_CLASS',\n",
    "                                     'ONCOTREE_CODE',\n",
    "                                     'CANCER_TYPE',\n",
    "                                     'CANCER_TYPE_DETAILED',\n",
    "                                     'TMB_NONSYNONYMOUS'\n",
    "                                    ], axis=1)\n",
    "\n",
    "\n",
    "print(sample_df.columns)\n",
    "\n",
    "\n",
    "# Check duplicate count\n",
    "num_duplicates = sample_df.duplicated().sum()\n",
    "print(f\"Number of duplicate rows : {num_duplicates}\")\n",
    "\n",
    "# print duplicates (excluding first instance)\n",
    "if num_duplicates > 0:\n",
    "    print(\"\\nDuplicate rows (excluding first instance):\")\n",
    "    print(sample_df[sample_df.duplicated()])\n",
    "\n",
    "# # print full duplicate groups (including the first occurrences)\n",
    "#     print(\"\\nAll rows involved in duplication:\")\n",
    "#     print(mut_df[mut_df.duplicated(keep=False)])\n",
    "\n",
    "# remove duplicates, but keep first occurrence\n",
    "    sample_df = sample_df.drop_duplicates()\n",
    "    print(f\"\\nDataFrame shape after removing duplicates: {sample_df.shape}\")\n",
    "else:\n",
    "    print(\"No duplicate rows found.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b671c4ad-d0b6-43e1-b15c-ec749665175d",
   "metadata": {},
   "source": [
    "#### Combine dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1766,
   "id": "ca4cd556-6160-44dd-9c56-f7f27a387f15",
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine dataframes\n",
    "\n",
    "init_combined_df = mut_df.merge(sample_df, on='SAMPLE_ID', how='left')\n",
    "\n",
    "# print(mut_df.columns)\n",
    "# print(mut_df.shape)\n",
    "\n",
    "# print(sample_df.columns)\n",
    "# print(sample_df.shape)\n",
    "\n",
    "# print(init_combined_df.columns)\n",
    "# print(init_combined_df.shape)\n",
    "\n",
    "#add patient_df\n",
    "combined_df = init_combined_df.merge(patient_df, on='PATIENT_ID', how='left')\n",
    "\n",
    "# print(patient_df.columns)\n",
    "# print(patient_df.shape)\n",
    "\n",
    "# print(combined_df.columns)\n",
    "# print(combined_df.shape)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1970da22-8dc7-4cc4-aa62-9becb291bbeb",
   "metadata": {},
   "source": [
    "#### Add STUDY_ID column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1767,
   "id": "0c2fd3ee-02fc-48bf-acc9-7675137eea80",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# add column for study id\n",
    "study_id = init_study_meta.iloc[0, 0]\n",
    "study_id = study_id.replace('cancer_study_identifier: ', '')\n",
    "# study_id\n",
    "combined_df['STUDY_ID'] = study_id\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90899975-c174-43d5-a7ae-2369d22f35f6",
   "metadata": {},
   "source": [
    "#### Check for (and remove) duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1768,
   "id": "aee6b500-7a61-4190-934b-53c7055686b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of duplicate rows : 0\n",
      "No duplicate rows found.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Check duplicate count\n",
    "num_duplicates = combined_df.duplicated().sum()\n",
    "print(f\"Number of duplicate rows : {num_duplicates}\")\n",
    "\n",
    "# print duplicates (excluding first instance)\n",
    "if num_duplicates > 0:\n",
    "    print(\"\\nDuplicate rows (excluding first instance):\")\n",
    "    print(combined_df[combined_df.duplicated()])\n",
    "\n",
    "# # print full duplicate groups (including the first occurrences)\n",
    "#     print(\"\\nAll rows involved in duplication:\")\n",
    "#     print(mut_df[mut_df.duplicated(keep=False)])\n",
    "\n",
    "# remove duplicates, but keep first occurrence\n",
    "    combined_df = combined_df.drop_duplicates()\n",
    "    print(f\"\\nDataFrame shape after removing duplicates: {combined_df.shape}\")\n",
    "else:\n",
    "    print(\"No duplicate rows found.\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da59990c-e8c3-41bd-ab79-4c0e505e1e72",
   "metadata": {},
   "source": [
    "#### Remove data from cell lines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1769,
   "id": "1b0bcca1-476e-4e77-a0a1-bb3a3bab9b1b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original shape: (14737, 27)\n",
      "Removed 3123 rows where SAMPLE_CLASS == 'Cell line'\n",
      "New shape: (11614, 27)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "SAMPLE_CLASS\n",
       "Cell line    3123\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 1769,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# remove cell lines\n",
    "\n",
    "original_shape = combined_df.shape\n",
    "print(f\"Original shape: {original_shape}\")\n",
    "\n",
    "#lines to remove\n",
    "removed_df = combined_df[combined_df['SAMPLE_CLASS'] == 'Cell line']\n",
    "\n",
    "# remove cell lines\n",
    "filtered_df = combined_df[combined_df['SAMPLE_CLASS'] != 'Cell line']\n",
    "\n",
    "# calculate how many rows were removed\n",
    "rows_removed = original_shape[0] - filtered_df.shape[0]\n",
    "print(f\"Removed {rows_removed} rows where SAMPLE_CLASS == 'Cell line'\")\n",
    "\n",
    "# print new shape\n",
    "print(f\"New shape: {filtered_df.shape}\")\n",
    "\n",
    "# reassign df\n",
    "combined_df = filtered_df\n",
    "\n",
    "removed_df.to_csv('cell_lines_removed.csv', index=False)\n",
    "removed_df.value_counts(\"SAMPLE_CLASS\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56b299dc-a8b9-4036-8431-7e336ea69170",
   "metadata": {},
   "source": [
    "#### Write value counts for NaNs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1770,
   "id": "96f130c2-8fd1-481b-8f5a-00db0ded9319",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Value counts written for 27 columns to folder: value_counts_by_column\n"
     ]
    }
   ],
   "source": [
    "# combined_df.isna().sum()\n",
    "\n",
    "# Hugo_Symbol                   0\n",
    "# Chromosome                    0\n",
    "# Start_Position                0\n",
    "# End_Position                  0\n",
    "# Consequence                 124 - some of the Variant_Classification=Silent have no consequence described\n",
    "# Variant_Classification        0\n",
    "# Variant_Type                  0\n",
    "# Reference_Allele              0\n",
    "# Tumor_Seq_Allele2             0\n",
    "# SAMPLE_ID                     0\n",
    "# Sequence_Source               0\n",
    "# HGVSc                       136 - some of the Variant_Classification=Silent and all of the Variant_Classification=3'Flank and 5'Flank\n",
    "# HGVSp                       346 \n",
    "# Transcript_ID               124 - some of the Variant_Classification=Silent have no Transcript_ID described\n",
    "# RefSeq                     1338\n",
    "# Protein_position            334\n",
    "# Gnomad_Notation               0\n",
    "# PATIENT_ID                    0\n",
    "# SAMPLE_CLASS                  0\n",
    "# ONCOTREE_CODE                 0\n",
    "# CANCER_TYPE                   0\n",
    "# CANCER_TYPE_DETAILED          0\n",
    "# TMB_NONSYNONYMOUS             0\n",
    "# AGE                        4745 - some ages undisclosed\n",
    "# SEX                           0\n",
    "# ETHNICITY                 10900 - many patients' ethnicities undisclosed\n",
    "# STUDY_ID                      0\n",
    "\n",
    "\n",
    "# define output folder (will create it if needed)\n",
    "output_dir = \"value_counts_by_column\"\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# loop through each column\n",
    "for col in combined_df.columns:\n",
    "    # build filename\n",
    "    filename = f\"value_counts_{col}.txt\"\n",
    "    filepath = os.path.join(output_dir, filename)\n",
    "    \n",
    "    # write counts to file\n",
    "    with open(filepath, \"w\") as f:\n",
    "        f.write(f\"Value counts for column: {col}\\n\\n\")\n",
    "        f.write(combined_df[col].value_counts(dropna=False).to_string())\n",
    "        f.write(\"\\n\")\n",
    "\n",
    "print(f\"✅ Value counts written for {len(combined_df.columns)} columns to folder: {output_dir}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24f92414-9951-4286-a145-adcc3dab3675",
   "metadata": {},
   "source": [
    "#### Replace missing patient, sample data with \"No_Data\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1771,
   "id": "5b032d44-fcdd-45c1-b4cb-6d42c6b23894",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hugo_Symbol                  0\n",
       "Chromosome                   0\n",
       "Start_Position               0\n",
       "End_Position                 0\n",
       "Consequence                  0\n",
       "Variant_Classification       0\n",
       "Variant_Type                 0\n",
       "Reference_Allele             0\n",
       "Tumor_Seq_Allele2            0\n",
       "SAMPLE_ID                    0\n",
       "Sequence_Source              0\n",
       "HGVSc                      136\n",
       "HGVSp                      346\n",
       "HGVSp_Short                136\n",
       "Transcript_ID              124\n",
       "RefSeq                    1338\n",
       "Protein_position           334\n",
       "PATIENT_ID                   0\n",
       "SAMPLE_CLASS                 0\n",
       "ONCOTREE_CODE                0\n",
       "CANCER_TYPE                  0\n",
       "CANCER_TYPE_DETAILED         0\n",
       "TMB_NONSYNONYMOUS            0\n",
       "AGE                          0\n",
       "SEX                          0\n",
       "ETHNICITY                    0\n",
       "STUDY_ID                     0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 1771,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#filling in NaNs - AGE, ETHNICITY, Consequence\n",
    "# TODO add consequence with annotation\n",
    "\n",
    "cols_to_fill = ['Consequence', 'AGE', 'ETHNICITY']\n",
    "fill_value = \"No_Data\"\n",
    "\n",
    "for col in cols_to_fill:\n",
    "    combined_df[col] = combined_df[col].fillna(fill_value)\n",
    "\n",
    "combined_df.isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ae16729-f4ba-44bf-bdbe-b312b46a8482",
   "metadata": {},
   "source": [
    "#### Construct TEMP GnomAD variant ID column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1772,
   "id": "2e34f131-d3b7-452c-9c3e-39f79b3157d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# construct Gnomad variant ID column\n",
    "combined_df[\"temp_Gnomad_Notation\"] = combined_df.apply(\n",
    "    lambda row: f\"{row['Chromosome']}-{row['Start_Position']}-{row['Reference_Allele']}-{row['Tumor_Seq_Allele2']}\",\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b71c762e-4548-439c-ba88-ec9dbfc049a5",
   "metadata": {},
   "source": [
    "#### Correcting Chromosome 23 samples to X or Y"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25862726-3cf2-41fc-91a6-ed9a0e6f9038",
   "metadata": {},
   "source": [
    "##### Write initial combined_df to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1773,
   "id": "14a83e5e-5984-4c3c-9d66-a87e7ebe31b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df.to_csv('output0.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b70f5960-4dfe-425e-8afb-64fc973af53c",
   "metadata": {},
   "source": [
    "##### Set test variables and REST API variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1774,
   "id": "91d4679e-bbff-4262-aa50-4c2cdfa574b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# BASE_URL = \"https://normalize.cancervariants.org/variation/\"\n",
    "# # HEADERS = {\"Accept\": \"application/json\"}\n",
    "# HEADERS = {\n",
    "#     \"Accept\": \"application/json\",\n",
    "#     \"User-Agent\": \"Mozilla/5.0 (Windows NT 10.0; Win64; x64)\"  # mimics a browser\n",
    "# }\n",
    "# variant = \"23-2408485-G-C\"\n",
    "# gene = \"ZBED1\"\n",
    "\n",
    "PATTERN = re.compile(r'^23-')        # anchored ^ so only the chromosome prefix is substituted\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1674efbf-645c-4fb0-8b5d-89aa0092ff16",
   "metadata": {},
   "source": [
    "##### FUNCTION: Flag rows with Chrom23"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1775,
   "id": "ae6af5f6-df95-407e-843b-5e500174b3f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def flag_rows_chrom_23(df):\n",
    "    \"\"\"\n",
    "    Create \"Chrom_23\" column, True for those with Chromosome = 23\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame\n",
    "        Must contain column 'Chromosome'.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    dataframe\n",
    "    \"\"\"\n",
    "    df[\"Chrom_23\"] = False\n",
    "    # print(combined_df.head)\n",
    "    df[\"Chrom_23\"] = df[\"Chromosome\"].astype(str).str.strip().eq(\"23\")\n",
    "    df.loc[df[\"Chromosome\"] == 23, \"Chrom_23\"] = True\n",
    "    # print(combined_df[\"Chrom_23\"].value_counts())\n",
    "    # print(combined_df[\"Chromosome\"].value_counts())\n",
    "    \n",
    "    return df\n",
    "\n",
    "combined_df = flag_rows_chrom_23(combined_df)\n",
    "\n",
    "combined_df.to_csv(\"output_flag_chrom_23.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adcedfb9-01e0-4b35-8b47-d174acc8913d",
   "metadata": {},
   "source": [
    "##### FUNCTION: change female chrom23 to X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1776,
   "id": "a8d93f16-6243-4659-a0f9-2dda39f41c45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(combined_df[\"Chromosome\"].value_counts())\n",
    "\n",
    "def chr23_female(df):\n",
    "    \"\"\"\n",
    "    Convert Chromosome 23 to 'X' for rows where SEX is female.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame\n",
    "        Must contain columns 'Chromosome' and 'SEX'.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    dataframe\n",
    "    \"\"\"\n",
    "    # Ensure we’re comparing like with like\n",
    "    chr_col = df[\"Chromosome\"].astype(str).str.strip()\n",
    "    sex_col = df[\"SEX\"].astype(str).str.upper().str.strip()   # handles 'F', 'f', 'Female', etc.\n",
    "    \n",
    "    mask = (chr_col == \"23\") & (sex_col.str.startswith(\"F\"))\n",
    "    df.loc[mask, \"Chromosome\"] = \"X\"\n",
    "    return df\n",
    "\n",
    "combined_df = chr23_female(combined_df)\n",
    "\n",
    "# print(combined_df[\"Chromosome\"].value_counts())\n",
    "\n",
    "combined_df.to_csv('output1_femaleX.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2561146-6f6b-42ab-80c1-02b741f144a6",
   "metadata": {},
   "source": [
    "##### FUNCTION: Add cols for Chr23_X and Chr23_Y, fill with false"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1777,
   "id": "1af4f568-4f1b-49db-b467-c96601462b46",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_cols_chrom_23_male(df):\n",
    "    \"\"\"\n",
    "    Create \"Chr23_X and Chr23_Y\" columns, fill with false\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame\n",
    "        Must contain column 'Chromosome'.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    dataframe\n",
    "    \"\"\"\n",
    "    df[\"Chr23_X\"] = False\n",
    "    df[\"Chr23_Y\"] = False\n",
    "    # df[\"Chrom_23\"] = df[\"Chromosome\"].astype(str).str.strip().eq(\"23\")\n",
    "    # df.loc[df[\"Chromosome\"] == 23, \"Chrom_23\"] = True\n",
    "    # print(combined_df[\"Chr23_X\"].value_counts())\n",
    "    # print(combined_df[\"Chr23_Y\"].value_counts())\n",
    "    return df\n",
    "\n",
    "combined_df = add_cols_chrom_23_male(combined_df)\n",
    "combined_df.to_csv('output2_new_chr23_boolean_cols.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b658bce-5d6f-4425-adae-c558405b5cd4",
   "metadata": {},
   "source": [
    "##### FUNCTION: Adjust GnomAD variant to accept X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1778,
   "id": "45a3f2f9-eb2f-436d-afb9-ba34549ab844",
   "metadata": {},
   "outputs": [],
   "source": [
    "def chr23_to_X(variant: str) -> str:\n",
    "    \"\"\"Convert a single '23-' prefix in a variant string to 'X-'.\"\"\"\n",
    "    return PATTERN.sub('X-', variant) if isinstance(variant, str) else variant\n",
    "\n",
    "x_variant = chr23_to_X(variant)\n",
    "# print(x_variant)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d3684d2-171a-4bec-a7e5-7e31fb981843",
   "metadata": {},
   "source": [
    "##### FUNCTION: Adjust GnomAD variant to accept Y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1779,
   "id": "3131ab23",
   "metadata": {},
   "outputs": [],
   "source": [
    "def chr23_to_Y(variant: str) -> str:\n",
    "    \"\"\"Convert a '23-' prefix in a single variant string to 'Y-'.\"\"\"\n",
    "    return PATTERN.sub('Y-', variant) if isinstance(variant, str) else variant\n",
    "\n",
    "y_variant = chr23_to_Y(variant)\n",
    "# print(y_variant)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29ff163d-467b-45f6-8fd8-f60c712b3218",
   "metadata": {},
   "source": [
    "##### FUNCTION: Test tokenization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1780,
   "id": "21497eb6-533b-4e1d-a205-0eb32d868be6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_variant_tokenization(variant: str, delay=0.5):\n",
    "    \"\"\"\n",
    "    Fetch normalized variant info from VICC API for a single variant string.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    variant : str\n",
    "        A GnomAD-style variant (e.g., '23-2408485-G-C').\n",
    "    delay : float\n",
    "        Seconds to wait between API requests (default 0.5).\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        DataFrame with original variant and raw JSON string response.\n",
    "    \"\"\"\n",
    "    \n",
    "    # BASE_URL = \"https://normalize.cancervariants.org/variation/\"\n",
    "    BASE_URL = \"http://localhost:8001/variation/\"\n",
    "    HEADERS = {\n",
    "        \"Accept\": \"application/json\",\n",
    "        \"User-Agent\": \"Mozilla/5.0 (Windows NT 10.0; Win64; x64)\"  # mimics a browser\n",
    "        }\n",
    "    \n",
    "    results = []\n",
    "\n",
    "    # url = f\"{BASE_URL}normalize?q={variant}\"\n",
    "    url = f\"{BASE_URL}normalize?q={variant}&hgvs_dup_del_mode=default&input_assembly=GRCh37\"\n",
    "\n",
    "    try:\n",
    "        response = requests.get(url, headers=HEADERS)\n",
    "        if response.status_code == 200:\n",
    "            data = response.json()\n",
    "            results.append({\n",
    "                \"variant\": variant,\n",
    "                \"response\": json.dumps(data)\n",
    "            })\n",
    "        else:\n",
    "            results.append({\n",
    "                \"variant\": variant,\n",
    "                \"response\": f\"Error {response.status_code}: {response.text}\"\n",
    "            })\n",
    "    except Exception as e:\n",
    "        results.append({\n",
    "            \"variant\": variant,\n",
    "            \"response\": f\"Exception: {str(e)}\"\n",
    "        })\n",
    "\n",
    "    time.sleep(delay)\n",
    "\n",
    "    return pd.DataFrame(results)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d447f008-e20c-4328-b44c-fd4c7589dabc",
   "metadata": {},
   "source": [
    "##### FUNCTION: Check if variant on X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1781,
   "id": "fb88a037-a608-45e5-91a2-1a71327e1d97",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_for_x_variant(df, variant):\n",
    "    # Convert the variant to X-style format (e.g., \"23-...\" → \"X-...\")\n",
    "    variant_x = chr23_to_X(variant)\n",
    "\n",
    "    # Query the API and get a one-row DataFrame\n",
    "    x_df = test_variant_tokenization(variant_x)  # returns a one-row DataFrame\n",
    "\n",
    "    # Extract and parse JSON string from the response\n",
    "    raw_response = x_df.loc[0, \"response\"]\n",
    "\n",
    "    try:\n",
    "        parsed_response = json.loads(raw_response)\n",
    "    except json.JSONDecodeError:\n",
    "        print(f\"❌ Failed to parse JSON for: {variant_x}\")\n",
    "        return df\n",
    "\n",
    "    # Ensure 'variation' is in the parsed response\n",
    "    if \"variation\" not in parsed_response:\n",
    "        print(f\"⚠️ 'variation' key not found for: {variant_x}\")\n",
    "        return df\n",
    "\n",
    "    # Try to extract hgnc_id if available\n",
    "    hgnc_id = None\n",
    "    try:\n",
    "        hgnc_id = parsed_response[\"variation\"][\"extensions\"][0][\"value\"][0][\"hgnc_id\"]\n",
    "        print(f\"✅ Extracted HGNC ID for {variant_x}: {hgnc_id}\")\n",
    "    except (KeyError, IndexError, TypeError):\n",
    "        print(f\"⚠️ No HGNC ID found for: {variant_x}\")\n",
    "\n",
    "    # Initialize columns if needed\n",
    "    if \"x_hgnc_id\" not in df.columns:\n",
    "        df[\"x_hgnc_id\"] = \"no_value\"\n",
    "    if \"Chr23_X\" not in df.columns:\n",
    "        df[\"Chr23_X\"] = False\n",
    "\n",
    "    # Reconstruct variant string from each row to match normalized variant_x\n",
    "    reconstructed = (\n",
    "        \"X-\" +\n",
    "        df[\"Start_Position\"].astype(str).str.strip() + \"-\" +\n",
    "        df[\"Reference_Allele\"].astype(str).str.strip() + \"-\" +\n",
    "        df[\"Tumor_Seq_Allele2\"].astype(str).str.strip()\n",
    "    )\n",
    "\n",
    "    # Mask: match rows on reconstructed X-variant and Chromosome == 23\n",
    "    chrom_col = df[\"Chromosome\"].astype(str).str.strip()\n",
    "    mask = (chrom_col == \"23\") & (reconstructed == variant_x)\n",
    "\n",
    "    print(f\"🧪 Matched {mask.sum()} row(s) for {variant_x}\")\n",
    "\n",
    "    # Set Chr23_X = True\n",
    "    df.loc[mask, \"Chr23_X\"] = True\n",
    "\n",
    "    # Set x_hgnc_id if available\n",
    "    if hgnc_id is not None:\n",
    "        df.loc[mask, \"x_hgnc_id\"] = hgnc_id\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90011ef4-8467-47e8-86a8-9e4c4f0b99c5",
   "metadata": {},
   "source": [
    "##### FUNCTION: Check if variant on Y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1782,
   "id": "1ce24b9c-aa33-4fef-a96c-43e168ddc3d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_for_y_variant(df, variant):\n",
    "    # Convert the variant to Y-style format (e.g., \"23-...\" → \"Y-...\")\n",
    "    variant_y = chr23_to_Y(variant)\n",
    "\n",
    "    # Query the API and get a one-row DataFrame\n",
    "    y_df = test_variant_tokenization(variant_y)  # returns a one-row DataFrame\n",
    "\n",
    "    # Extract and parse JSON string from the response\n",
    "    raw_response = y_df.loc[0, \"response\"]\n",
    "\n",
    "    try:\n",
    "        parsed_response = json.loads(raw_response)\n",
    "    except json.JSONDecodeError:\n",
    "        print(f\"❌ Failed to parse JSON for: {variant_y}\")\n",
    "        return df\n",
    "\n",
    "    # Ensure 'variation' is in the parsed response\n",
    "    if \"variation\" not in parsed_response:\n",
    "        print(f\"⚠️ 'variation' key not found for: {variant_y}\")\n",
    "        return df\n",
    "\n",
    "    # Try to extract hgnc_id if available\n",
    "    hgnc_id = None\n",
    "    try:\n",
    "        hgnc_id = parsed_response[\"variation\"][\"extensions\"][0][\"value\"][0][\"hgnc_id\"]\n",
    "        print(f\"✅ Extracted HGNC ID for {variant_y}: {hgnc_id}\")\n",
    "    except (KeyError, IndexError, TypeError):\n",
    "        print(f\"⚠️ No HGNC ID found for: {variant_y}\")\n",
    "\n",
    "    # Initialize columns if needed\n",
    "    if \"y_hgnc_id\" not in df.columns:\n",
    "        df[\"y_hgnc_id\"] = \"no_value\"\n",
    "    if \"Chr23_Y\" not in df.columns:\n",
    "        df[\"Chr23_Y\"] = False\n",
    "\n",
    "    # Reconstruct variant string from each row to match normalized variant_y\n",
    "    reconstructed = (\n",
    "        \"Y-\" +\n",
    "        df[\"Start_Position\"].astype(str).str.strip() + \"-\" +\n",
    "        df[\"Reference_Allele\"].astype(str).str.strip() + \"-\" +\n",
    "        df[\"Tumor_Seq_Allele2\"].astype(str).str.strip()\n",
    "    )\n",
    "\n",
    "    # Mask: match rows on reconstructed Y-variant and Chromosome == 23\n",
    "    chrom_col = df[\"Chromosome\"].astype(str).str.strip()\n",
    "    mask = (chrom_col == \"23\") & (reconstructed == variant_y)\n",
    "\n",
    "    print(f\"🧪 Matched {mask.sum()} row(s) for {variant_y}\")\n",
    "\n",
    "    # Set Chr23_Y = True\n",
    "    df.loc[mask, \"Chr23_Y\"] = True\n",
    "\n",
    "    # Set y_hgnc_id if available\n",
    "    if hgnc_id is not None:\n",
    "        df.loc[mask, \"y_hgnc_id\"] = hgnc_id\n",
    "\n",
    "    return df\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18252a87-c901-46f8-9b13-715d20f32099",
   "metadata": {},
   "source": [
    "##### FUNCTION: Master function for dealing with male chrom23 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1783,
   "id": "0a18fa41-7ef2-49a9-a4f4-ffb183fb881d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def chr23_male(df, variant):\n",
    "    # df = add_cols_chrom_23_male(df)          # prep chromosome-23 columns\n",
    "    if \"Chr23_X\" not in df.columns:\n",
    "        df[\"Chr23_X\"] = False\n",
    "    if \"Chr23_Y\" not in df.columns:\n",
    "        df[\"Chr23_Y\"] = False\n",
    "    df = check_for_x_variant(df, variant)    # pass **both** args\n",
    "    df = check_for_y_variant(df, variant)\n",
    "\n",
    "    df.to_csv(\"output2_new_chr23_boolean_cols.csv\", index=False)\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "787a27a5",
   "metadata": {},
   "source": [
    "##### Setting chromosome 23 variant notations as a list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1784,
   "id": "a7f214e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total variants in list: 53\n",
      "Sample of chrom_23_list: ['23-70824012-C-T', '23-84258927-C-T', '23-153184693-C-T', '23-153175777-G-A', '23-135772859-A-C']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "chrom_23_list = combined_df[\n",
    "    (combined_df[\"SEX\"].str.lower() == \"male\") &\n",
    "    (combined_df[\"Chrom_23\"] == True)\n",
    "][\"temp_Gnomad_Notation\"].dropna().tolist()\n",
    "\n",
    "print(\"Total variants in list:\", len(chrom_23_list))\n",
    "print(\"Sample of chrom_23_list:\", chrom_23_list[:5])\n",
    "\n",
    "# print(len(chrom_23_list))\n",
    "\n",
    "# with open(\"chrom_23_list.json\", \"w\") as f:\n",
    "#     json.dump(chrom_23_list, f, indent=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48852d18",
   "metadata": {},
   "source": [
    "##### Initialize x_hgnc_id and y_hgnc_id columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1785,
   "id": "7d61f8f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"x_hgnc_id\" not in combined_df.columns:\n",
    "    result_df[\"x_hgnc_id\"] = \"no_value\"\n",
    "if \"y_hgnc_id\" not in combined_df.columns:\n",
    "    result_df[\"y_hgnc_id\"] = \"no_value\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86247e94-f6e5-4036-b60f-77f9ea327693",
   "metadata": {},
   "source": [
    "##### RUN: driver function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1786,
   "id": "91705e89-0bb1-43e4-acfb-e65ee791c519",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "▶️ Checking variant: 23-70824012-C-T\n",
      "✅ Extracted HGNC ID for X-70824012-C-T: 15805\n",
      "🧪 Matched 1 row(s) for X-70824012-C-T\n",
      "⚠️ 'variation' key not found for: Y-70824012-C-T\n",
      "▶️ Checking variant: 23-84258927-C-T\n",
      "✅ Extracted HGNC ID for X-84258927-C-T: 24009\n",
      "🧪 Matched 1 row(s) for X-84258927-C-T\n",
      "⚠️ 'variation' key not found for: Y-84258927-C-T\n",
      "▶️ Checking variant: 23-153184693-C-T\n",
      "✅ Extracted HGNC ID for X-153184693-C-T: 674\n",
      "🧪 Matched 1 row(s) for X-153184693-C-T\n",
      "⚠️ 'variation' key not found for: Y-153184693-C-T\n",
      "▶️ Checking variant: 23-153175777-G-A\n",
      "✅ Extracted HGNC ID for X-153175777-G-A: 674\n",
      "🧪 Matched 1 row(s) for X-153175777-G-A\n",
      "⚠️ 'variation' key not found for: Y-153175777-G-A\n",
      "▶️ Checking variant: 23-135772859-A-C\n",
      "✅ Extracted HGNC ID for X-135772859-A-C: 685\n",
      "🧪 Matched 1 row(s) for X-135772859-A-C\n",
      "⚠️ 'variation' key not found for: Y-135772859-A-C\n",
      "▶️ Checking variant: 23-135827412-A-G\n",
      "✅ Extracted HGNC ID for X-135827412-A-G: 685\n",
      "🧪 Matched 1 row(s) for X-135827412-A-G\n",
      "⚠️ 'variation' key not found for: Y-135827412-A-G\n",
      "▶️ Checking variant: 23-18192265-T-A\n",
      "✅ Extracted HGNC ID for X-18192265-T-A: 28509\n",
      "🧪 Matched 1 row(s) for X-18192265-T-A\n",
      "⚠️ 'variation' key not found for: Y-18192265-T-A\n",
      "▶️ Checking variant: 23-15552431-G-A\n",
      "✅ Extracted HGNC ID for X-15552431-G-A: 1079\n",
      "🧪 Matched 1 row(s) for X-15552431-G-A\n",
      "⚠️ 'variation' key not found for: Y-15552431-G-A\n",
      "▶️ Checking variant: 23-1428413-C-T\n",
      "✅ Extracted HGNC ID for X-1428413-C-T: 2435\n",
      "🧪 Matched 1 row(s) for X-1428413-C-T\n",
      "⚠️ No HGNC ID found for: Y-1428413-C-T\n",
      "🧪 Matched 1 row(s) for Y-1428413-C-T\n",
      "▶️ Checking variant: 23-32519921-C-T\n",
      "✅ Extracted HGNC ID for X-32519921-C-T: 2928\n",
      "🧪 Matched 1 row(s) for X-32519921-C-T\n",
      "⚠️ 'variation' key not found for: Y-32519921-C-T\n",
      "▶️ Checking variant: 23-71428259-A-G\n",
      "✅ Extracted HGNC ID for X-71428259-A-G: 20794\n",
      "🧪 Matched 1 row(s) for X-71428259-A-G\n",
      "⚠️ 'variation' key not found for: Y-71428259-A-G\n",
      "▶️ Checking variant: 23-70444064-C-T\n",
      "✅ Extracted HGNC ID for X-70444064-C-T: 4283\n",
      "🧪 Matched 1 row(s) for X-70444064-C-T\n",
      "⚠️ 'variation' key not found for: Y-70444064-C-T\n",
      "▶️ Checking variant: 23-108718554-T-C\n",
      "✅ Extracted HGNC ID for X-108718554-T-C: 4691\n",
      "🧪 Matched 1 row(s) for X-108718554-T-C\n",
      "⚠️ 'variation' key not found for: Y-108718554-T-C\n",
      "▶️ Checking variant: 23-152734633-G-C\n",
      "✅ Extracted HGNC ID for X-152734633-G-C: 32979\n",
      "🧪 Matched 1 row(s) for X-152734633-G-C\n",
      "⚠️ 'variation' key not found for: Y-152734633-G-C\n",
      "▶️ Checking variant: 23-153223671-C-T\n",
      "✅ Extracted HGNC ID for X-153223671-C-T: 4839\n",
      "🧪 Matched 1 row(s) for X-153223671-C-T\n",
      "⚠️ 'variation' key not found for: Y-153223671-C-T\n",
      "▶️ Checking variant: 23-53575184-C-T\n",
      "✅ Extracted HGNC ID for X-53575184-C-T: 30892\n",
      "🧪 Matched 1 row(s) for X-53575184-C-T\n",
      "⚠️ 'variation' key not found for: Y-53575184-C-T\n",
      "▶️ Checking variant: 23-53247029-G-A\n",
      "✅ Extracted HGNC ID for X-53247029-G-A: 11114\n",
      "🧪 Matched 1 row(s) for X-53247029-G-A\n",
      "⚠️ 'variation' key not found for: Y-53247029-G-A\n",
      "▶️ Checking variant: 23-30268970-A-G\n",
      "✅ Extracted HGNC ID for X-30268970-A-G: 6808\n",
      "🧪 Matched 1 row(s) for X-30268970-A-G\n",
      "⚠️ 'variation' key not found for: Y-30268970-A-G\n",
      "▶️ Checking variant: 23-30237495-T-C\n",
      "✅ Extracted HGNC ID for X-30237495-T-C: 6809\n",
      "🧪 Matched 1 row(s) for X-30237495-T-C\n",
      "⚠️ 'variation' key not found for: Y-30237495-T-C\n",
      "▶️ Checking variant: 23-30237111-G-A\n",
      "✅ Extracted HGNC ID for X-30237111-G-A: 6809\n",
      "🧪 Matched 1 row(s) for X-30237111-G-A\n",
      "⚠️ 'variation' key not found for: Y-30237111-G-A\n",
      "▶️ Checking variant: 23-19391803-G-A\n",
      "✅ Extracted HGNC ID for X-19391803-G-A: 31689\n",
      "🧪 Matched 1 row(s) for X-19391803-G-A\n",
      "⚠️ 'variation' key not found for: Y-19391803-G-A\n",
      "▶️ Checking variant: 23-70357689-C-T\n",
      "✅ Extracted HGNC ID for X-70357689-C-T: 11957\n",
      "🧪 Matched 1 row(s) for X-70357689-C-T\n",
      "⚠️ 'variation' key not found for: Y-70357689-C-T\n",
      "▶️ Checking variant: 23-106186330-C-T\n",
      "✅ Extracted HGNC ID for X-106186330-C-T: 23485\n",
      "🧪 Matched 1 row(s) for X-106186330-C-T\n",
      "⚠️ 'variation' key not found for: Y-106186330-C-T\n",
      "▶️ Checking variant: 23-118725211-A-G\n",
      "⚠️ No HGNC ID found for: X-118725211-A-G\n",
      "🧪 Matched 1 row(s) for X-118725211-A-G\n",
      "⚠️ 'variation' key not found for: Y-118725211-A-G\n",
      "▶️ Checking variant: 23-46918083-T-C\n",
      "✅ Extracted HGNC ID for X-46918083-T-C: 22982\n",
      "🧪 Matched 1 row(s) for X-46918083-T-C\n",
      "⚠️ 'variation' key not found for: Y-46918083-T-C\n",
      "▶️ Checking variant: 23-152160017-G-C\n",
      "✅ Extracted HGNC ID for X-152160017-G-C: 18743\n",
      "🧪 Matched 1 row(s) for X-152160017-G-C\n",
      "⚠️ 'variation' key not found for: Y-152160017-G-C\n",
      "▶️ Checking variant: 23-47030516-C-T\n",
      "✅ Extracted HGNC ID for X-47030516-C-T: 9896\n",
      "🧪 Matched 1 row(s) for X-47030516-C-T\n",
      "⚠️ 'variation' key not found for: Y-47030516-C-T\n",
      "▶️ Checking variant: 23-153209091-G-A\n",
      "✅ Extracted HGNC ID for X-153209091-G-A: 9959\n",
      "🧪 Matched 1 row(s) for X-153209091-G-A\n",
      "⚠️ 'variation' key not found for: Y-153209091-G-A\n",
      "▶️ Checking variant: 23-109697532-A-G\n",
      "✅ Extracted HGNC ID for X-109697532-A-G: 29245\n",
      "🧪 Matched 1 row(s) for X-109697532-A-G\n",
      "⚠️ 'variation' key not found for: Y-109697532-A-G\n",
      "▶️ Checking variant: 23-106144104-G-A\n",
      "✅ Extracted HGNC ID for X-106144104-G-A: 25117\n",
      "🧪 Matched 1 row(s) for X-106144104-G-A\n",
      "⚠️ 'variation' key not found for: Y-106144104-G-A\n",
      "▶️ Checking variant: 23-73740919-C-T\n",
      "✅ Extracted HGNC ID for X-73740919-C-T: 10923\n",
      "🧪 Matched 1 row(s) for X-73740919-C-T\n",
      "⚠️ 'variation' key not found for: Y-73740919-C-T\n",
      "▶️ Checking variant: 23-129499602-C-T\n",
      "✅ Extracted HGNC ID for X-129499602-C-T: 10984\n",
      "🧪 Matched 1 row(s) for X-129499602-C-T\n",
      "⚠️ 'variation' key not found for: Y-129499602-C-T\n",
      "▶️ Checking variant: 23-135104837-A-G\n",
      "✅ Extracted HGNC ID for X-135104837-A-G: 11079\n",
      "🧪 Matched 1 row(s) for X-135104837-A-G\n",
      "⚠️ 'variation' key not found for: Y-135104837-A-G\n",
      "▶️ Checking variant: 23-21995227-C-T\n",
      "✅ Extracted HGNC ID for X-21995227-C-T: 11123\n",
      "🧪 Matched 1 row(s) for X-21995227-C-T\n",
      "⚠️ 'variation' key not found for: Y-21995227-C-T\n",
      "▶️ Checking variant: 23-155003998-C-T\n",
      "✅ Extracted HGNC ID for X-155003998-C-T: 11271\n",
      "🧪 Matched 1 row(s) for X-155003998-C-T\n",
      "⚠️ 'variation' key not found for: Y-155003998-C-T\n",
      "▶️ Checking variant: 23-48116682-C-T\n",
      "✅ Extracted HGNC ID for X-48116682-C-T: 11335\n",
      "🧪 Matched 1 row(s) for X-48116682-C-T\n",
      "⚠️ 'variation' key not found for: Y-48116682-C-T\n",
      "▶️ Checking variant: 23-30849637-G-A\n",
      "✅ Extracted HGNC ID for X-30849637-G-A: 30681\n",
      "🧪 Matched 1 row(s) for X-30849637-G-A\n",
      "⚠️ 'variation' key not found for: Y-30849637-G-A\n",
      "▶️ Checking variant: 23-79286496-T-C\n",
      "✅ Extracted HGNC ID for X-79286496-T-C: 11600\n",
      "🧪 Matched 1 row(s) for X-79286496-T-C\n",
      "⚠️ 'variation' key not found for: Y-79286496-T-C\n",
      "▶️ Checking variant: 23-101395767-C-T\n",
      "⚠️ No HGNC ID found for: X-101395767-C-T\n",
      "🧪 Matched 1 row(s) for X-101395767-C-T\n",
      "⚠️ 'variation' key not found for: Y-101395767-C-T\n",
      "▶️ Checking variant: 23-107018344-C-T\n",
      "✅ Extracted HGNC ID for X-107018344-C-T: 3051\n",
      "🧪 Matched 1 row(s) for X-107018344-C-T\n",
      "⚠️ 'variation' key not found for: Y-107018344-C-T\n",
      "▶️ Checking variant: 23-53115005-C-T\n",
      "✅ Extracted HGNC ID for X-53115005-C-T: 24358\n",
      "🧪 Matched 1 row(s) for X-53115005-C-T\n",
      "⚠️ 'variation' key not found for: Y-53115005-C-T\n",
      "▶️ Checking variant: 23-47104792-C-T\n",
      "✅ Extracted HGNC ID for X-47104792-C-T: 12609\n",
      "🧪 Matched 1 row(s) for X-47104792-C-T\n",
      "⚠️ 'variation' key not found for: Y-47104792-C-T\n",
      "▶️ Checking variant: 23-40990737-G-A\n",
      "✅ Extracted HGNC ID for X-40990737-G-A: 12632\n",
      "🧪 Matched 1 row(s) for X-40990737-G-A\n",
      "⚠️ 'variation' key not found for: Y-40990737-G-A\n",
      "▶️ Checking variant: 23-155171659-A-C\n",
      "✅ Extracted HGNC ID for X-155171659-A-C: 11486\n",
      "🧪 Matched 1 row(s) for X-155171659-A-C\n",
      "⚠️ 'variation' key not found for: Y-155171659-A-C\n",
      "▶️ Checking variant: 23-7812030-C-T\n",
      "✅ Extracted HGNC ID for X-7812030-C-T: 12667\n",
      "🧪 Matched 1 row(s) for X-7812030-C-T\n",
      "⚠️ 'variation' key not found for: Y-7812030-C-T\n",
      "▶️ Checking variant: 23-8434397-C-T\n",
      "✅ Extracted HGNC ID for X-8434397-C-T: 31838\n",
      "🧪 Matched 1 row(s) for X-8434397-C-T\n",
      "⚠️ No HGNC ID found for: Y-8434397-C-T\n",
      "🧪 Matched 1 row(s) for Y-8434397-C-T\n",
      "▶️ Checking variant: 23-10096159-C-T\n",
      "⚠️ No HGNC ID found for: X-10096159-C-T\n",
      "🧪 Matched 1 row(s) for X-10096159-C-T\n",
      "⚠️ 'variation' key not found for: Y-10096159-C-T\n",
      "▶️ Checking variant: 23-128876154-C-T\n",
      "✅ Extracted HGNC ID for X-128876154-C-T: 12823\n",
      "🧪 Matched 1 row(s) for X-128876154-C-T\n",
      "⚠️ 'variation' key not found for: Y-128876154-C-T\n",
      "▶️ Checking variant: 23-2408485-G-C\n",
      "✅ Extracted HGNC ID for X-2408485-G-C: 447\n",
      "🧪 Matched 1 row(s) for X-2408485-G-C\n",
      "⚠️ 'variation' key not found for: Y-2408485-G-C\n",
      "▶️ Checking variant: 23-101138860-T-C\n",
      "✅ Extracted HGNC ID for X-101138860-T-C: 29377\n",
      "🧪 Matched 1 row(s) for X-101138860-T-C\n",
      "⚠️ 'variation' key not found for: Y-101138860-T-C\n",
      "▶️ Checking variant: 23-152089287-G-A\n",
      "✅ Extracted HGNC ID for X-152089287-G-A: 12976\n",
      "🧪 Matched 1 row(s) for X-152089287-G-A\n",
      "⚠️ 'variation' key not found for: Y-152089287-G-A\n",
      "▶️ Checking variant: 23-129370462-G-A\n",
      "✅ Extracted HGNC ID for X-129370462-G-A: 25955\n",
      "🧪 Matched 1 row(s) for X-129370462-G-A\n",
      "⚠️ 'variation' key not found for: Y-129370462-G-A\n",
      "▶️ Checking variant: 23-134494503-C-T\n",
      "✅ Extracted HGNC ID for X-134494503-C-T: 21039\n",
      "🧪 Matched 1 row(s) for X-134494503-C-T\n",
      "⚠️ 'variation' key not found for: Y-134494503-C-T\n",
      "Chr23_X\n",
      "False    11561\n",
      "True        53\n",
      "Name: count, dtype: int64\n",
      "Chr23_Y\n",
      "False    11612\n",
      "True         2\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Start with a copy of your combined_df\n",
    "result_df = combined_df.copy()\n",
    "\n",
    "# Iterate through each chromosome 23 variant for male samples\n",
    "for variant in chrom_23_list:\n",
    "    print(f\"▶️ Checking variant: {variant}\")  # Add this line\n",
    "    result_df = chr23_male(result_df, variant)\n",
    "\n",
    "# Now all updates are preserved in result_df\n",
    "print(result_df[\"Chr23_X\"].value_counts(dropna=False))\n",
    "print(result_df[\"Chr23_Y\"].value_counts(dropna=False))\n",
    "\n",
    "# Save the final output to CSV\n",
    "result_df.to_csv(\"output_post_23_BOOLEAN.csv\", index=False)\n",
    "\n",
    "##TODO: THIS IS ONLY CHANGING ZNF449 TO TRUE FOR X\n",
    "##TODO: SOME GENES SHOW ALIASES, SOME ARE WRONG GENES\n",
    "##TODO: MUST TRUST THAT NORMALIZER IS DOING ITS JOB. WHEN A NEW SYMBOL COMES BACK, BUT THEY ARE ALIASES, NORMALIZER GIVES \n",
    "##TODO: USE GENEID INSTEAD OF GENE SYMBOL FOR TOKENIZATION TEST. DO VARIATION TESTER FOR GENE SYMBOL AND THEN GET GENE ID. THEN TEST VARIANT, GET GENE ID. IF THEY ARE THE SAME, RECORD.  \n",
    "##TODO: why aren't some of the x variants showing up with hgnc IDs.  Like X-84258927-C-T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9573318-5a06-4f42-a58a-9bc1a6f56ef9",
   "metadata": {},
   "source": [
    "##### FUNCTION: Reassign male chrom 23s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1787,
   "id": "de80033f-138b-4cd3-ba04-ce074cc61768",
   "metadata": {},
   "outputs": [],
   "source": [
    "def correct_male_chrom23(df):\n",
    "    # Initialize ambig_chrom column\n",
    "    df[\"ambig_chrom\"] = \"non-ambiguous\"\n",
    "\n",
    "    def update_row(row):\n",
    "        if row.get(\"Chrom_23\") is True and str(row.get(\"SEX\", \"\")).strip().lower() == \"male\":\n",
    "            if row[\"Chr23_X\"] and not row[\"Chr23_Y\"]:\n",
    "                row[\"Chromosome\"] = \"X\"\n",
    "            elif row[\"Chr23_Y\"] and not row[\"Chr23_X\"]:\n",
    "                row[\"Chromosome\"] = \"Y\"\n",
    "            elif row[\"Chr23_X\"] and row[\"Chr23_Y\"]:\n",
    "                row[\"ambig_chrom\"] = \"XY\"\n",
    "            else:  # neither Chr23_X nor Chr23_Y is True\n",
    "                row[\"ambig_chrom\"] = \"neither\"\n",
    "        return row\n",
    "\n",
    "    # Apply corrections row-by-row\n",
    "    df = df.apply(update_row, axis=1)\n",
    "\n",
    "    # Check for ambiguous rows\n",
    "    ambig_rows = df[df[\"ambig_chrom\"].isin([\"XY\", \"neither\"])]\n",
    "    if not ambig_rows.empty:\n",
    "        print(f\"⚠️ Found {len(ambig_rows)} row(s) with ambiguous chromosome identification.\")\n",
    "        print(ambig_rows[[\"temp_Gnomad_Notation\", \"Hugo_Symbol\", \"ambig_chrom\"]].head())\n",
    "\n",
    "    return df\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c09cbb2-bd23-4424-87f4-05cb9f21ae61",
   "metadata": {},
   "source": [
    "##### RUN: Correct male 23s and see if any ambiguous chromosome 23 values exist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1788,
   "id": "6dbe5829-d125-46b7-9a56-4411b305aec9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "⚠️ Found 2 row(s) with ambiguous chromosome identification.\n",
      "      temp_Gnomad_Notation Hugo_Symbol ambig_chrom\n",
      "2906        23-1428413-C-T      CSF2RA          XY\n",
      "10778       23-8434397-C-T       VCX3B          XY\n",
      "ambig_chrom\n",
      "non-ambiguous    11612\n",
      "XY                   2\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "result_df = correct_male_chrom23(result_df)\n",
    "\n",
    "print(result_df[\"ambig_chrom\"].value_counts())\n",
    "\n",
    "result_df.to_csv(\"output_post_23_correction.csv\", index=False)\n",
    "\n",
    "# result_df \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1d91b95",
   "metadata": {},
   "source": [
    "##### FUNCTION: correct ambiguous chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1789,
   "id": "8c8c0df7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def resolve_ambiguous_chromosomes(df):\n",
    "    \"\"\"\n",
    "    Resolve ambiguous Chr23 variants in male samples and flag problematic rows.\n",
    "\n",
    "    - For Chrom_23 == True and SEX == male:\n",
    "      - If both x_hgnc_id and y_hgnc_id have values → ⚠️ warn\n",
    "      - If neither has values → ❌ warn\n",
    "      - If ambig_chrom == \"XY\" or \"neither\", try to resolve Chromosome field\n",
    "    \"\"\"\n",
    "    chrom23_male_mask = (\n",
    "        (df[\"Chrom_23\"] == True) &\n",
    "        (df[\"SEX\"].str.lower().str.strip() == \"male\")\n",
    "    )\n",
    "\n",
    "    for idx, row in df[chrom23_male_mask].iterrows():\n",
    "        x_id = str(row.get(\"x_hgnc_id\", \"\")).strip()\n",
    "        y_id = str(row.get(\"y_hgnc_id\", \"\")).strip()\n",
    "        ambig_status = row.get(\"ambig_chrom\", \"not_set\")\n",
    "\n",
    "        # ⚠️ Raise general warnings, regardless of ambig_chrom\n",
    "        if x_id != \"no_value\" and y_id != \"no_value\":\n",
    "            print(f\"⚠️ Row {idx}: Both x_hgnc_id and y_hgnc_id have values. Manual check recommended.\")\n",
    "        elif x_id == \"no_value\" and y_id == \"no_value\":\n",
    "            print(f\"❌ Row {idx}: Neither x_hgnc_id nor y_hgnc_id has a value.\")\n",
    "\n",
    "        # ✅ Try resolving ambiguous chromosomes\n",
    "        if ambig_status in [\"XY\", \"neither\"]:\n",
    "            if x_id != \"no_value\" and y_id == \"no_value\":\n",
    "                df.at[idx, \"Chromosome\"] = \"X\"\n",
    "                df.at[idx, \"ambig_chrom\"] = \"resolved_X\"\n",
    "                print(f\"✅ Resolved index {idx} to Chromosome X\")\n",
    "            elif y_id != \"no_value\" and x_id == \"no_value\":\n",
    "                df.at[idx, \"Chromosome\"] = \"Y\"\n",
    "                df.at[idx, \"ambig_chrom\"] = \"resolved_Y\"\n",
    "                print(f\"✅ Resolved index {idx} to Chromosome Y\")\n",
    "            elif x_id != \"no_value\" and y_id != \"no_value\":\n",
    "                print(f\"⚠️ Index {idx}: Ambiguous Chromosome with both HGNC IDs present.\")\n",
    "            else:\n",
    "                print(f\"❌ Index {idx}: Ambiguous Chromosome with no HGNC ID found.\")\n",
    "\n",
    "    return df\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acf98c7f",
   "metadata": {},
   "source": [
    "##### RUN: resolve ambiguous chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1790,
   "id": "a04899f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Resolved index 2906 to Chromosome X\n",
      "❌ Row 6805: Neither x_hgnc_id nor y_hgnc_id has a value.\n",
      "❌ Row 9877: Neither x_hgnc_id nor y_hgnc_id has a value.\n",
      "✅ Resolved index 10778 to Chromosome X\n",
      "❌ Row 10942: Neither x_hgnc_id nor y_hgnc_id has a value.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>SAMPLE_ID</th>\n",
       "      <th>...</th>\n",
       "      <th>SEX</th>\n",
       "      <th>ETHNICITY</th>\n",
       "      <th>STUDY_ID</th>\n",
       "      <th>temp_Gnomad_Notation</th>\n",
       "      <th>Chrom_23</th>\n",
       "      <th>Chr23_X</th>\n",
       "      <th>Chr23_Y</th>\n",
       "      <th>x_hgnc_id</th>\n",
       "      <th>y_hgnc_id</th>\n",
       "      <th>ambig_chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ETFDH</td>\n",
       "      <td>4</td>\n",
       "      <td>159603535</td>\n",
       "      <td>159603535</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SJDES004</td>\n",
       "      <td>...</td>\n",
       "      <td>Male</td>\n",
       "      <td>White/Europe</td>\n",
       "      <td>es_dfarber_broad_2014</td>\n",
       "      <td>4-159603535-G-C</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no_value</td>\n",
       "      <td>no_value</td>\n",
       "      <td>non-ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GJB4</td>\n",
       "      <td>1</td>\n",
       "      <td>35227334</td>\n",
       "      <td>35227334</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SJDES004</td>\n",
       "      <td>...</td>\n",
       "      <td>Male</td>\n",
       "      <td>White/Europe</td>\n",
       "      <td>es_dfarber_broad_2014</td>\n",
       "      <td>1-35227334-G-A</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no_value</td>\n",
       "      <td>no_value</td>\n",
       "      <td>non-ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MUC5B</td>\n",
       "      <td>11</td>\n",
       "      <td>1272679</td>\n",
       "      <td>1272679</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SJDES004</td>\n",
       "      <td>...</td>\n",
       "      <td>Male</td>\n",
       "      <td>White/Europe</td>\n",
       "      <td>es_dfarber_broad_2014</td>\n",
       "      <td>11-1272679-G-A</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no_value</td>\n",
       "      <td>no_value</td>\n",
       "      <td>non-ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PCDHAC1</td>\n",
       "      <td>5</td>\n",
       "      <td>140307515</td>\n",
       "      <td>140307515</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SJDES004</td>\n",
       "      <td>...</td>\n",
       "      <td>Male</td>\n",
       "      <td>White/Europe</td>\n",
       "      <td>es_dfarber_broad_2014</td>\n",
       "      <td>5-140307515-G-A</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no_value</td>\n",
       "      <td>no_value</td>\n",
       "      <td>non-ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YAP1</td>\n",
       "      <td>11</td>\n",
       "      <td>102080254</td>\n",
       "      <td>102080254</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SJDES004</td>\n",
       "      <td>...</td>\n",
       "      <td>Male</td>\n",
       "      <td>White/Europe</td>\n",
       "      <td>es_dfarber_broad_2014</td>\n",
       "      <td>11-102080254-C-T</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no_value</td>\n",
       "      <td>no_value</td>\n",
       "      <td>non-ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14667</th>\n",
       "      <td>ZNF41</td>\n",
       "      <td>X</td>\n",
       "      <td>47306909</td>\n",
       "      <td>47306909</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>CHEWS017</td>\n",
       "      <td>...</td>\n",
       "      <td>Male</td>\n",
       "      <td>No_Data</td>\n",
       "      <td>es_dfarber_broad_2014</td>\n",
       "      <td>X-47306909-T-C</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no_value</td>\n",
       "      <td>no_value</td>\n",
       "      <td>non-ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14668</th>\n",
       "      <td>ZNF645</td>\n",
       "      <td>X</td>\n",
       "      <td>22292037</td>\n",
       "      <td>22292037</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CHEWS002</td>\n",
       "      <td>...</td>\n",
       "      <td>Male</td>\n",
       "      <td>No_Data</td>\n",
       "      <td>es_dfarber_broad_2014</td>\n",
       "      <td>X-22292037-G-A</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no_value</td>\n",
       "      <td>no_value</td>\n",
       "      <td>non-ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14669</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>X</td>\n",
       "      <td>15841117</td>\n",
       "      <td>15841117</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>01-P034</td>\n",
       "      <td>...</td>\n",
       "      <td>Male</td>\n",
       "      <td>No_Data</td>\n",
       "      <td>es_dfarber_broad_2014</td>\n",
       "      <td>X-15841117-C-T</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no_value</td>\n",
       "      <td>no_value</td>\n",
       "      <td>non-ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14670</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>X</td>\n",
       "      <td>15841120</td>\n",
       "      <td>15841120</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>01-P131</td>\n",
       "      <td>...</td>\n",
       "      <td>Female</td>\n",
       "      <td>No_Data</td>\n",
       "      <td>es_dfarber_broad_2014</td>\n",
       "      <td>X-15841120-C-T</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no_value</td>\n",
       "      <td>no_value</td>\n",
       "      <td>non-ambiguous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14671</th>\n",
       "      <td>ZXDA</td>\n",
       "      <td>X</td>\n",
       "      <td>57935484</td>\n",
       "      <td>57935484</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>CHEWS036</td>\n",
       "      <td>...</td>\n",
       "      <td>Male</td>\n",
       "      <td>No_Data</td>\n",
       "      <td>es_dfarber_broad_2014</td>\n",
       "      <td>X-57935484-A-T</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no_value</td>\n",
       "      <td>no_value</td>\n",
       "      <td>non-ambiguous</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11614 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0           ETFDH          4       159603535     159603535   \n",
       "1            GJB4          1        35227334      35227334   \n",
       "2           MUC5B         11         1272679       1272679   \n",
       "3         PCDHAC1          5       140307515     140307515   \n",
       "4            YAP1         11       102080254     102080254   \n",
       "...           ...        ...             ...           ...   \n",
       "14667       ZNF41          X        47306909      47306909   \n",
       "14668      ZNF645          X        22292037      22292037   \n",
       "14669       ZRSR2          X        15841117      15841117   \n",
       "14670       ZRSR2          X        15841120      15841120   \n",
       "14671        ZXDA          X        57935484      57935484   \n",
       "\n",
       "              Consequence Variant_Classification Variant_Type  \\\n",
       "0        missense_variant      Missense_Mutation          SNP   \n",
       "1        missense_variant      Missense_Mutation          SNP   \n",
       "2        missense_variant      Missense_Mutation          SNP   \n",
       "3      synonymous_variant                 Silent          SNP   \n",
       "4        missense_variant      Missense_Mutation          SNP   \n",
       "...                   ...                    ...          ...   \n",
       "14667    missense_variant      Missense_Mutation          SNP   \n",
       "14668    missense_variant      Missense_Mutation          SNP   \n",
       "14669    missense_variant      Missense_Mutation          SNP   \n",
       "14670    missense_variant      Missense_Mutation          SNP   \n",
       "14671         stop_gained      Nonsense_Mutation          SNP   \n",
       "\n",
       "      Reference_Allele Tumor_Seq_Allele2 SAMPLE_ID  ...     SEX     ETHNICITY  \\\n",
       "0                    G                 C  SJDES004  ...    Male  White/Europe   \n",
       "1                    G                 A  SJDES004  ...    Male  White/Europe   \n",
       "2                    G                 A  SJDES004  ...    Male  White/Europe   \n",
       "3                    G                 A  SJDES004  ...    Male  White/Europe   \n",
       "4                    C                 T  SJDES004  ...    Male  White/Europe   \n",
       "...                ...               ...       ...  ...     ...           ...   \n",
       "14667                T                 C  CHEWS017  ...    Male       No_Data   \n",
       "14668                G                 A  CHEWS002  ...    Male       No_Data   \n",
       "14669                C                 T   01-P034  ...    Male       No_Data   \n",
       "14670                C                 T   01-P131  ...  Female       No_Data   \n",
       "14671                A                 T  CHEWS036  ...    Male       No_Data   \n",
       "\n",
       "                    STUDY_ID temp_Gnomad_Notation Chrom_23 Chr23_X  Chr23_Y  \\\n",
       "0      es_dfarber_broad_2014      4-159603535-G-C    False   False    False   \n",
       "1      es_dfarber_broad_2014       1-35227334-G-A    False   False    False   \n",
       "2      es_dfarber_broad_2014       11-1272679-G-A    False   False    False   \n",
       "3      es_dfarber_broad_2014      5-140307515-G-A    False   False    False   \n",
       "4      es_dfarber_broad_2014     11-102080254-C-T    False   False    False   \n",
       "...                      ...                  ...      ...     ...      ...   \n",
       "14667  es_dfarber_broad_2014       X-47306909-T-C    False   False    False   \n",
       "14668  es_dfarber_broad_2014       X-22292037-G-A    False   False    False   \n",
       "14669  es_dfarber_broad_2014       X-15841117-C-T    False   False    False   \n",
       "14670  es_dfarber_broad_2014       X-15841120-C-T    False   False    False   \n",
       "14671  es_dfarber_broad_2014       X-57935484-A-T    False   False    False   \n",
       "\n",
       "      x_hgnc_id y_hgnc_id    ambig_chrom  \n",
       "0      no_value  no_value  non-ambiguous  \n",
       "1      no_value  no_value  non-ambiguous  \n",
       "2      no_value  no_value  non-ambiguous  \n",
       "3      no_value  no_value  non-ambiguous  \n",
       "4      no_value  no_value  non-ambiguous  \n",
       "...         ...       ...            ...  \n",
       "14667  no_value  no_value  non-ambiguous  \n",
       "14668  no_value  no_value  non-ambiguous  \n",
       "14669  no_value  no_value  non-ambiguous  \n",
       "14670  no_value  no_value  non-ambiguous  \n",
       "14671  no_value  no_value  non-ambiguous  \n",
       "\n",
       "[11614 rows x 34 columns]"
      ]
     },
     "execution_count": 1790,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resolve_ambiguous_chromosomes(result_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42f33921",
   "metadata": {},
   "source": [
    "##### Create Gnomad_Notation column and populate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1791,
   "id": "e79566c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# construct Gnomad variant ID column\n",
    "result_df[\"Gnomad_Notation\"] = result_df.apply(\n",
    "    lambda row: f\"{row['Chromosome']}-{row['Start_Position']}-{row['Reference_Allele']}-{row['Tumor_Seq_Allele2']}\",\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02ba77c8",
   "metadata": {},
   "source": [
    "##### FUNCTION: test gene tokenization\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1792,
   "id": "e6a190ae-bf47-45bc-814e-b4c3ec41875c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_gene_tokenization(gene: str, delay=0.5):\n",
    "    BASE_URL = \"https://normalize.cancervariants.org/gene/\"\n",
    "    HEADERS = {\n",
    "        \"Accept\": \"application/json\",\n",
    "        \"User-Agent\": \"Mozilla/5.0\"\n",
    "    }\n",
    "\n",
    "    results = []\n",
    "\n",
    "    url = f\"{BASE_URL}normalize?q={gene}\"\n",
    "\n",
    "    try:\n",
    "        response = requests.get(url, headers=HEADERS)\n",
    "        if response.status_code == 200:\n",
    "            data = response.json()\n",
    "            results.append({\n",
    "                \"gene\": gene,\n",
    "                \"response\": json.dumps(data)\n",
    "            })\n",
    "        else:\n",
    "            results.append({\n",
    "                \"gene\": gene,\n",
    "                \"response\": f\"Error {response.status_code}: {response.text}\"\n",
    "            })\n",
    "    except Exception as e:\n",
    "        results.append({\n",
    "            \"gene\": gene,\n",
    "            \"response\": f\"Exception: {str(e)}\"\n",
    "        })\n",
    "\n",
    "    time.sleep(delay)\n",
    "    return pd.DataFrame(results)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7748ae6b",
   "metadata": {},
   "source": [
    "##### FUNCTION: test gene tokenization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e020666",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def update_gene_hgnc_id_inplace(df):\n",
    "    for idx, row in df.iterrows():\n",
    "        if row.get(\"Chrom_23\") and str(row.get(\"SEX\", \"\")).strip().lower() == \"male\":\n",
    "            try:\n",
    "                result = test_gene_tokenization(row[\"Hugo_Symbol\"])\n",
    "                full_id = result[\"gene\"][\"id\"]  # e.g., \"normalize.gene.hgnc:447\"\n",
    "                hgnc_id = full_id.split(\":\")[-1]  # Extract \"447\"\n",
    "                df.at[idx, \"gene_hgnc_id\"] = hgnc_id\n",
    "            except Exception as e:\n",
    "                print(f\"⚠️ Could not extract HGNC ID for {row['Hugo_Symbol']}: {e}\")\n",
    "\n",
    "    print(df.head())\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f14ec1c2",
   "metadata": {},
   "source": [
    "##### Setting genes as list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1805,
   "id": "6de02e23",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total genes in list: 53\n",
      "Sample of gene_list: ['ACRC', 'APOOL', 'ARHGAP4', 'ARHGAP4', 'ARHGEF6']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Ensure column exists and is initialized to \"untested\"\n",
    "if \"gene_hgnc_id\" not in result_df.columns:\n",
    "    result_df[\"gene_hgnc_id\"] = \"untested\"\n",
    "\n",
    "gene_list = result_df[\n",
    "    (result_df[\"SEX\"].str.lower() == \"male\") &\n",
    "    (result_df[\"Chrom_23\"] == True)\n",
    "][\"Hugo_Symbol\"].dropna().tolist()\n",
    "\n",
    "print(\"Total genes in list:\", len(gene_list))\n",
    "print(\"Sample of gene_list:\", gene_list[:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3ed87ec",
   "metadata": {},
   "source": [
    "##### RUN: Gene tokenization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1806,
   "id": "6be61b2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "▶️ Checking gene: ACRC\n",
      "✅ Extracted HGNC ID for ACRC: 15805\n",
      "▶️ Checking gene: APOOL\n",
      "✅ Extracted HGNC ID for APOOL: 24009\n",
      "▶️ Checking gene: ARHGAP4\n",
      "✅ Extracted HGNC ID for ARHGAP4: 674\n",
      "▶️ Checking gene: ARHGAP4\n",
      "✅ Extracted HGNC ID for ARHGAP4: 674\n",
      "▶️ Checking gene: ARHGEF6\n",
      "✅ Extracted HGNC ID for ARHGEF6: 685\n",
      "▶️ Checking gene: ARHGEF6\n",
      "✅ Extracted HGNC ID for ARHGEF6: 685\n",
      "▶️ Checking gene: BEND2\n",
      "✅ Extracted HGNC ID for BEND2: 28509\n",
      "▶️ Checking gene: BMX\n",
      "✅ Extracted HGNC ID for BMX: 1079\n",
      "▶️ Checking gene: CSF2RA\n",
      "✅ Extracted HGNC ID for CSF2RA: 2435\n",
      "▶️ Checking gene: DMD\n",
      "✅ Extracted HGNC ID for DMD: 2928\n",
      "▶️ Checking gene: ERCC6L\n",
      "✅ Extracted HGNC ID for ERCC6L: 20794\n",
      "▶️ Checking gene: GJB1\n",
      "✅ Extracted HGNC ID for GJB1: 4283\n",
      "▶️ Checking gene: GUCY2F\n",
      "✅ Extracted HGNC ID for GUCY2F: 4691\n",
      "▶️ Checking gene: HAUS7\n",
      "✅ Extracted HGNC ID for HAUS7: 32979\n",
      "▶️ Checking gene: HCFC1\n",
      "✅ Extracted HGNC ID for HCFC1: 4839\n",
      "▶️ Checking gene: HUWE1\n",
      "✅ Extracted HGNC ID for HUWE1: 30892\n",
      "▶️ Checking gene: KDM5C\n",
      "✅ Extracted HGNC ID for KDM5C: 11114\n",
      "▶️ Checking gene: MAGEB1\n",
      "✅ Extracted HGNC ID for MAGEB1: 6808\n",
      "▶️ Checking gene: MAGEB2\n",
      "✅ Extracted HGNC ID for MAGEB2: 6809\n",
      "▶️ Checking gene: MAGEB2\n",
      "✅ Extracted HGNC ID for MAGEB2: 6809\n",
      "▶️ Checking gene: MAP3K15\n",
      "✅ Extracted HGNC ID for MAP3K15: 31689\n",
      "▶️ Checking gene: MED12\n",
      "✅ Extracted HGNC ID for MED12: 11957\n",
      "▶️ Checking gene: MORC4\n",
      "✅ Extracted HGNC ID for MORC4: 23485\n",
      "▶️ Checking gene: NKRF\n",
      "✅ Extracted HGNC ID for NKRF: 19374\n",
      "▶️ Checking gene: PHF16\n",
      "✅ Extracted HGNC ID for PHF16: 22982\n",
      "▶️ Checking gene: PNMA5\n",
      "✅ Extracted HGNC ID for PNMA5: 18743\n",
      "▶️ Checking gene: RBM10\n",
      "✅ Extracted HGNC ID for RBM10: 9896\n",
      "▶️ Checking gene: RENBP\n",
      "✅ Extracted HGNC ID for RENBP: 9959\n",
      "▶️ Checking gene: RGAG1\n",
      "✅ Extracted HGNC ID for RGAG1: 29245\n",
      "▶️ Checking gene: RIPPLY1\n",
      "✅ Extracted HGNC ID for RIPPLY1: 25117\n",
      "▶️ Checking gene: SLC16A2\n",
      "✅ Extracted HGNC ID for SLC16A2: 10923\n",
      "▶️ Checking gene: SLC25A14\n",
      "✅ Extracted HGNC ID for SLC25A14: 10984\n",
      "▶️ Checking gene: SLC9A6\n",
      "✅ Extracted HGNC ID for SLC9A6: 11079\n",
      "▶️ Checking gene: SMS\n",
      "✅ Extracted HGNC ID for SMS: 11123\n",
      "▶️ Checking gene: SPRY3\n",
      "✅ Extracted HGNC ID for SPRY3: 11271\n",
      "▶️ Checking gene: SSX1\n",
      "✅ Extracted HGNC ID for SSX1: 11335\n",
      "▶️ Checking gene: TAB3\n",
      "✅ Extracted HGNC ID for TAB3: 30681\n",
      "▶️ Checking gene: TBX22\n",
      "✅ Extracted HGNC ID for TBX22: 11600\n",
      "▶️ Checking gene: TCEAL6\n",
      "✅ Extracted HGNC ID for TCEAL6: 24553\n",
      "▶️ Checking gene: TSC22D3\n",
      "✅ Extracted HGNC ID for TSC22D3: 3051\n",
      "▶️ Checking gene: TSPYL2\n",
      "✅ Extracted HGNC ID for TSPYL2: 24358\n",
      "▶️ Checking gene: USP11\n",
      "✅ Extracted HGNC ID for USP11: 12609\n",
      "▶️ Checking gene: USP9X\n",
      "✅ Extracted HGNC ID for USP9X: 12632\n",
      "▶️ Checking gene: VAMP7\n",
      "✅ Extracted HGNC ID for VAMP7: 11486\n",
      "▶️ Checking gene: VCX\n",
      "✅ Extracted HGNC ID for VCX: 12667\n",
      "▶️ Checking gene: VCX3B\n",
      "✅ Extracted HGNC ID for VCX3B: 31838\n",
      "▶️ Checking gene: WWC3\n",
      "✅ Extracted HGNC ID for WWC3: 29237\n",
      "▶️ Checking gene: XPNPEP2\n",
      "✅ Extracted HGNC ID for XPNPEP2: 12823\n",
      "▶️ Checking gene: ZBED1\n",
      "✅ Extracted HGNC ID for ZBED1: 447\n",
      "▶️ Checking gene: ZMAT1\n",
      "✅ Extracted HGNC ID for ZMAT1: 29377\n",
      "▶️ Checking gene: ZNF185\n",
      "✅ Extracted HGNC ID for ZNF185: 12976\n",
      "▶️ Checking gene: ZNF280C\n",
      "✅ Extracted HGNC ID for ZNF280C: 25955\n",
      "▶️ Checking gene: ZNF449\n",
      "✅ Extracted HGNC ID for ZNF449: 21039\n",
      "gene_hgnc_id\n",
      "untested    11561\n",
      "674             2\n",
      "685             2\n",
      "6809            2\n",
      "24358           1\n",
      "10984           1\n",
      "11079           1\n",
      "11123           1\n",
      "11271           1\n",
      "11335           1\n",
      "30681           1\n",
      "11600           1\n",
      "24553           1\n",
      "3051            1\n",
      "12632           1\n",
      "12609           1\n",
      "25117           1\n",
      "11486           1\n",
      "12667           1\n",
      "31838           1\n",
      "29237           1\n",
      "12823           1\n",
      "447             1\n",
      "29377           1\n",
      "12976           1\n",
      "25955           1\n",
      "10923           1\n",
      "9959            1\n",
      "29245           1\n",
      "4839            1\n",
      "24009           1\n",
      "28509           1\n",
      "1079            1\n",
      "2435            1\n",
      "2928            1\n",
      "20794           1\n",
      "4283            1\n",
      "4691            1\n",
      "32979           1\n",
      "30892           1\n",
      "15805           1\n",
      "11114           1\n",
      "6808            1\n",
      "31689           1\n",
      "11957           1\n",
      "23485           1\n",
      "19374           1\n",
      "22982           1\n",
      "18743           1\n",
      "9896            1\n",
      "21039           1\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Start with a copy of your combined_df\n",
    "result_post_23_df = result_df.copy()\n",
    "\n",
    "def populate_gene_hgnc_col(gene_list, df):\n",
    "    # Add the column once\n",
    "    if \"gene_hgnc_id\" not in df.columns:\n",
    "        df[\"gene_hgnc_id\"] = None\n",
    "\n",
    "    for gene in gene_list:\n",
    "        print(f\"▶️ Checking gene: {gene}\")\n",
    "        gene_df = test_gene_tokenization(gene)\n",
    "\n",
    "        # Extract and parse JSON string from the response\n",
    "        raw_response = gene_df.loc[0, \"response\"]\n",
    "        try:\n",
    "            parsed_response = json.loads(raw_response)\n",
    "        except json.JSONDecodeError:\n",
    "            print(f\"❌ Failed to parse JSON for: {gene}\")\n",
    "            continue\n",
    "\n",
    "        if \"gene\" not in parsed_response:\n",
    "            print(f\"⚠️ 'gene' key not found for: {gene}\")\n",
    "            continue\n",
    "\n",
    "        try:\n",
    "            hgnc_id = parsed_response[\"gene\"][\"id\"].split(\":\")[-1]\n",
    "            print(f\"✅ Extracted HGNC ID for {gene}: {hgnc_id}\")\n",
    "        except (KeyError, IndexError, TypeError):\n",
    "            print(f\"⚠️ No HGNC ID found for: {gene}\")\n",
    "            continue\n",
    "\n",
    "        # Apply to matching rows\n",
    "        df.loc[\n",
    "            (df[\"Chrom_23\"] == True) & \n",
    "            (df[\"SEX\"].str.strip().str.lower() == \"male\") & \n",
    "            (df[\"Hugo_Symbol\"].str.strip() == gene.strip()),\n",
    "            \"gene_hgnc_id\"\n",
    "        ] = hgnc_id\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "result_post_23_df = populate_gene_hgnc_col(gene_list, result_post_23_df)\n",
    "print(result_post_23_df[\"gene_hgnc_id\"].value_counts(dropna=False))\n",
    "result_post_23_df.to_csv(\"output_post_gene_norm.csv\", index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5d370f1",
   "metadata": {},
   "source": [
    "##### FUNCTION: check hgnc_id matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1807,
   "id": "3179ad7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def validate_gene_hgnc_match(df):\n",
    "    \"\"\"\n",
    "    Compare gene_hgnc_id to x_hgnc_id or y_hgnc_id for Chromosome 23 variants in male samples.\n",
    "    \n",
    "    Adds a column 'hgnc_id_match':\n",
    "        - \"untested\" (default)\n",
    "        - set to matching hgnc_id if match is found\n",
    "        - \"no_match\" if mismatch is found\n",
    "\n",
    "    Prints warnings for mismatches.\n",
    "    \"\"\"\n",
    "    # Initialize column\n",
    "    if \"hgnc_id_match\" not in df.columns:\n",
    "        df[\"hgnc_id_match\"] = \"untested\"\n",
    "\n",
    "    # Define mask for male Chr23 variants\n",
    "    mask = (df[\"Chrom_23\"] == True) & (df[\"SEX\"].str.strip().str.lower() == \"male\")\n",
    "\n",
    "    for idx, row in df[mask].iterrows():\n",
    "        gene_id = str(row.get(\"gene_hgnc_id\", \"\")).strip()\n",
    "        chrom = str(row.get(\"Chromosome\", \"\")).strip()\n",
    "        x_id = str(row.get(\"x_hgnc_id\", \"\")).strip()\n",
    "        y_id = str(row.get(\"y_hgnc_id\", \"\")).strip()\n",
    "\n",
    "        if chrom == \"X\":\n",
    "            if gene_id == x_id and gene_id != \"no_value\":\n",
    "                df.at[idx, \"hgnc_id_match\"] = gene_id\n",
    "            else:\n",
    "                df.at[idx, \"hgnc_id_match\"] = \"no_match\"\n",
    "                print(f\"❌ Row {idx}: X chromosome mismatch. gene_hgnc_id={gene_id}, x_hgnc_id={x_id}\")\n",
    "        elif chrom == \"Y\":\n",
    "            if gene_id == y_id and gene_id != \"no_value\":\n",
    "                df.at[idx, \"hgnc_id_match\"] = gene_id\n",
    "            else:\n",
    "                df.at[idx, \"hgnc_id_match\"] = \"no_match\"\n",
    "                print(f\"❌ Row {idx}: Y chromosome mismatch. gene_hgnc_id={gene_id}, y_hgnc_id={y_id}\")\n",
    "\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4760c2ab",
   "metadata": {},
   "source": [
    "##### RUN: validate hgnc ID matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1808,
   "id": "bf5ad738",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "❌ Row 6805: X chromosome mismatch. gene_hgnc_id=19374, x_hgnc_id=no_value\n",
      "❌ Row 9877: X chromosome mismatch. gene_hgnc_id=24553, x_hgnc_id=no_value\n",
      "❌ Row 10942: X chromosome mismatch. gene_hgnc_id=29237, x_hgnc_id=no_value\n"
     ]
    }
   ],
   "source": [
    "post_validation_df = validate_gene_hgnc_match(result_post_23_df)\n",
    "post_validation_df.to_csv(\"output_post_validation.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8e13d06",
   "metadata": {},
   "source": [
    "#### Function: fill in missing values in df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e122f03",
   "metadata": {},
   "source": [
    "##### FUNCTION: Query API for HGVSp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1811,
   "id": "0659b463",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>response</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>X-70824012-C-T</td>\n",
       "      <td>{\"warnings\": [], \"service_meta_\": {\"name\": \"va...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          variant                                           response\n",
       "0  X-70824012-C-T  {\"warnings\": [], \"service_meta_\": {\"name\": \"va..."
      ]
     },
     "execution_count": 1811,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check NAs \n",
    "# post_validation_df.isna().sum()\n",
    "variant = \"X-70824012-C-T\"\n",
    "def retrieve_HGVSp_value(variant: str, delay=0.5):\n",
    "    BASE_URL = \"http://localhost:8001/variation/\"\n",
    "    HEADERS = {\n",
    "        \"Accept\": \"application/json\",\n",
    "        \"User-Agent\": \"Mozilla/5.0\"\n",
    "    }\n",
    "\n",
    "    results = []\n",
    "\n",
    "    url = f\"{BASE_URL}gnomad_vcf_to_protein?q={variant}\"\n",
    "\n",
    "    try:\n",
    "        response = requests.get(url, headers=HEADERS)\n",
    "        if response.status_code == 200:\n",
    "            data = response.json()\n",
    "            results.append({\n",
    "                \"variant\": variant,\n",
    "                \"response\": json.dumps(data)\n",
    "            })\n",
    "        else:\n",
    "            results.append({\n",
    "                \"variant\": variant,\n",
    "                \"response\": f\"Error {response.status_code}: {response.text}\"\n",
    "            })\n",
    "    except Exception as e:\n",
    "        results.append({\n",
    "            \"variant\": variant,\n",
    "            \"response\": f\"Exception: {str(e)}\"\n",
    "        })\n",
    "\n",
    "    time.sleep(delay)\n",
    "    return pd.DataFrame(results)\n",
    "\n",
    "r = retrieve_HGVSp_value(variant)\n",
    "r\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31b808ea",
   "metadata": {},
   "source": [
    "##### FUNCTION: update HGVSp in place for missing values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00105f39",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def update_HGVSp_inplace(df):\n",
    "    for idx, row in df.iterrows():\n",
    "        if row.get(\"Chrom_23\") and str(row.get(\"SEX\", \"\")).strip().lower() == \"male\":\n",
    "            try:\n",
    "                result = test_gene_tokenization(row[\"Hugo_Symbol\"])\n",
    "                full_id = result[\"gene\"][\"id\"]  # e.g., \"normalize.gene.hgnc:447\"\n",
    "                hgnc_id = full_id.split(\":\")[-1]  # Extract \"447\"\n",
    "                df.at[idx, \"gene_hgnc_id\"] = hgnc_id\n",
    "            except Exception as e:\n",
    "                print(f\"⚠️ Could not extract HGNC ID for {row['Hugo_Symbol']}: {e}\")\n",
    "\n",
    "    print(df.head())\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc34dd19-40d4-4505-9775-f35062b52820",
   "metadata": {},
   "source": [
    "##### Next step?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee2f0126-563b-424e-b26e-f27d3a909d59",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "#remove temporary columns\n",
    "\n",
    "#get missing p dots!\n",
    "\n",
    "#consolidate driver function\n",
    "\n",
    "#Need function that tkes gnomad notation and puts it through normalizer, then takes response and fashions it into the test fixture\n",
    "#put response into columns \"focus variant members\", etc. \n",
    "\n",
    "#figure out other info for object - cohort, frequency, etc. \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 515,
   "id": "d56ef311",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "Index(['Gnomad_Notation'], dtype='object')",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/n7/jxfygfkj2w182jgqtmnmzl7r0000gq/T/ipykernel_33058/3686122230.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# remove variant dupes per patient\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;31m# find duplicated (PATIENT_ID, Gnomad_Notation) pairs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mdupe_mask\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcombined_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mduplicated\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubset\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"PATIENT_ID\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"Gnomad_Notation\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"first\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0;31m# new DataFrame with the duplicated rows\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0mpatient_variant_dupes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcombined_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdupe_mask\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;31m# remove those rows from the original DataFrame\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/repos/metakb/venv/lib/python3.11/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, subset, keep)\u001b[0m\n\u001b[1;32m   6946\u001b[0m         \u001b[0;31m# Otherwise, raise a KeyError, same as if you try to __getitem__ with a\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6947\u001b[0m         \u001b[0;31m# key that doesn't exist.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6948\u001b[0m         \u001b[0mdiff\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubset\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6949\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mdiff\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6950\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mIndex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdiff\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6951\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6952\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubset\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m1\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_unique\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6953\u001b[0m             \u001b[0;31m# GH#45236 This is faster than get_group_index below\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: Index(['Gnomad_Notation'], dtype='object')"
     ]
    }
   ],
   "source": [
    "# remove variant dupes per patient\n",
    "\n",
    "# find duplicated (PATIENT_ID, Gnomad_Notation) pairs\n",
    "dupe_mask = combined_df.duplicated(subset=[\"PATIENT_ID\", \"Gnomad_Notation\"], keep=\"first\")\n",
    "# new DataFrame with the duplicated rows\n",
    "patient_variant_dupes = combined_df[dupe_mask]\n",
    "# remove those rows from the original DataFrame\n",
    "combined_df_cleaned = combined_df[~dupe_mask]\n",
    "# write removed rows to file\n",
    "patient_variant_dupes.to_csv(\"patient_variant_dupes.csv\", index=False)\n",
    "# print the number of rows removed\n",
    "print(f\"Removed {patient_variant_dupes.shape[0]} rows with duplicated Gnomad_Notation per PATIENT_ID.\")\n",
    "# reassign dataframe:\n",
    "combined_df = combined_df_cleaned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b3a1e72",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6f5738f-f282-4550-9c76-d1b3c82f3880",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
