{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Graph database for storing harvested data.\"\"\"\n",
    "from neo4j import GraphDatabase\n",
    "from neo4j.exceptions import ServiceUnavailable\n",
    "from typing import Tuple, Dict, List\n",
    "import logging\n",
    "import json\n",
    "\n",
    "logger = logging.getLogger('metakb')\n",
    "logger.setLevel(logging.DEBUG)\n",
    "\n",
    "\n",
    "class Graph:\n",
    "    \"\"\"Manage requests to graph datastore.\n",
    "\n",
    "    TODO\n",
    "    * Need to add (currently excluding):\n",
    "       * Expressions in Variation Descriptors\n",
    "       * Extensions in any Descriptors (json serialize?)\n",
    "          * including Variant Groups\n",
    "       * any Assertion objects\n",
    "    * format the long constaints string better\n",
    "    * do any current serialized values need to be searchable?\n",
    "    * refactor add_transformed_data()\n",
    "    * do evidence line/assertion need specific methods?\n",
    "    * handle other types of propositions?\n",
    "    * refactor the ugly key/value property string\n",
    "    * can Labels be retrieved as part of the query?\n",
    "    * handle prime (') symbols in strings -- skipping for now (see whitelist)\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, uri: str, credentials: Tuple[str, str]):\n",
    "        \"\"\"Initialize Graph driver instance.\n",
    "        :param str uri: address of Neo4j DB\n",
    "        :param Tuple[str, str] credentials: tuple containing username and\n",
    "            password\n",
    "        \"\"\"\n",
    "        self.driver = GraphDatabase.driver(uri, auth=credentials)\n",
    "        with self.driver.session() as session:\n",
    "            session.write_transaction(self._create_constraints)\n",
    "\n",
    "    def close(self):\n",
    "        \"\"\"Close Neo4j driver.\"\"\"\n",
    "        self.driver.close()\n",
    "\n",
    "    def clear(self):\n",
    "        \"\"\"Debugging helper - wipe out DB.\"\"\"\n",
    "        def delete_all(tx):\n",
    "            tx.run(\"MATCH (n) DETACH DELETE n;\")\n",
    "        with self.driver.session() as session:\n",
    "            session.write_transaction(delete_all)\n",
    "\n",
    "    @staticmethod\n",
    "    def _create_constraints(tx):\n",
    "        \"\"\"Create unique property constraints for ID values\"\"\"\n",
    "        try:\n",
    "            tx.run(\"CREATE CONSTRAINT gene_id_constraint IF NOT EXISTS ON (n:Gene) ASSERT n.gene_id IS UNIQUE;\")  # noqa: E501\n",
    "            tx.run(\"CREATE CONSTRAINT disease_id_constraint IF NOT EXISTS ON (n:Disease) ASSERT n.disease_id IS UNIQUE;\")  # noqa: E501\n",
    "            tx.run(\"CREATE CONSTRAINT therapy_id_constraint IF NOT EXISTS ON (n:Therapy) ASSERT n.therapy_id IS UNIQUE;\")  # noqa: E501\n",
    "            tx.run(\"CREATE CONSTRAINT vod_id_constraint IF NOT EXISTS ON (n:ValueObjectDescriptor) ASSERT n.id IS UNIQUE;\")  # noqa: E501\n",
    "            tx.run(\"CREATE CONSTRAINT proposition_id_constraint IF NOT EXISTS ON (n:Proposition) ASSERT n._id IS UNIQUE;\")  # noqa: E501\n",
    "            tx.run(\"CREATE CONSTRAINT evidence_id_constraint IF NOT EXISTS ON (n:Evidence) ASSERT n.id IS UNIQUE;\")  # noqa: E501\n",
    "            tx.run(\"CREATE CONSTRAINT assertionmethod_id_constraint IF NOT EXISTS ON (n:AssertionMethod) ASSERT n.id IS UNIQUE;\")  # noqa: E501\n",
    "        except ServiceUnavailable as exception:\n",
    "            logging.error(\"Failed to generate ID property constraints.\")\n",
    "            raise exception\n",
    "\n",
    "    def add_transformed_data(self, data: Dict):\n",
    "        \"\"\"Add set of data formatted per Common Data Model to DB.\n",
    "        :param Dict data: contains Evidence, Proposition, Assertion Methods,\n",
    "            and Values/Descriptors for Gene, Disease, Therapy, and Variation\n",
    "        \"\"\"\n",
    "        with self.driver.session() as session:\n",
    "            for proposition in data.get('propositions', []):\n",
    "                session.write_transaction(self._add_therapeutic_response,\n",
    "                                          proposition)\n",
    "            for method in data.get('assertion_methods', []):\n",
    "                session.write_transaction(self._add_assertion_method, method)\n",
    "            for descr in data.get('therapy_descriptors', []):\n",
    "                session.write_transaction(self._add_descriptor, descr)\n",
    "            for descr in data.get('disease_descriptors', []):\n",
    "                session.write_transaction(self._add_descriptor, descr)\n",
    "            for descr in data.get('gene_descriptors', []):\n",
    "                session.write_transaction(self._add_descriptor, descr)\n",
    "            for var_descr in data.get('variation_descriptors', []):\n",
    "                session.write_transaction(self._add_variant_descriptor,\n",
    "                                          var_descr)\n",
    "            for ev in data.get('evidence', []):\n",
    "                session.write_transaction(self._add_evidence, ev)\n",
    "\n",
    "    @staticmethod\n",
    "    def _add_therapeutic_response(tx, ther_response: Dict):\n",
    "        \"\"\"Add Therapeutic Response object to DB.\n",
    "        :param Dict ther_response: TherapeuticResponse object as dict\n",
    "        \"\"\"\n",
    "        nonnull_values = [f'{key}:\"{ther_response[key]}\"' for key \n",
    "                          in ('_id', 'predicate', 'variation_origin')\n",
    "                          if ther_response[key]]\n",
    "        values_joined = ', '.join(nonnull_values)\n",
    "        query = f\"\"\"\n",
    "        MERGE (n:TherapeuticResponse:Proposition\n",
    "            {{ {values_joined} }});\n",
    "        \"\"\"\n",
    "        try:\n",
    "            tx.run(query, **ther_response)\n",
    "        except ServiceUnavailable as exception:\n",
    "            logging.error(f\"Failed to add TherapeuticResponse object\\n\"\n",
    "                          f\"Query: {query}\\nTherapeuticResponse: \"\n",
    "                          f\"{ther_response}\")\n",
    "            raise exception\n",
    "\n",
    "    @staticmethod\n",
    "    def _add_assertion_method(tx, assertion_method: Dict):\n",
    "        \"\"\"Add Assertion Method object to DB.\"\"\"\n",
    "        assertion_method['version'] = \"20091210\"\n",
    "        query = \"\"\"\n",
    "        MERGE (n:AssertionMethod {id:$id, label:$label, url:$url,\n",
    "            version:$version, reference: $reference});\n",
    "        \"\"\"\n",
    "        try:\n",
    "            tx.run(query, **assertion_method)\n",
    "        except ServiceUnavailable as exception:\n",
    "            logging.error(f\"Failed to add AssertionMethod object\\nQuery: \"\n",
    "                          f\"{query}\\nAssertionMethod: {assertion_method}\")\n",
    "            raise exception\n",
    "\n",
    "    @staticmethod\n",
    "    def _add_descriptor(tx, descriptor: Dict):\n",
    "        \"\"\"Add descriptor object to DB.\"\"\"\n",
    "        descr_type = descriptor['type']\n",
    "        if descr_type == 'TherapyDescriptor':\n",
    "            value_type = 'Therapy'\n",
    "            value_id_name = 'therapy_id'\n",
    "            value_id = descriptor['value']['therapy_id']\n",
    "        elif descr_type == 'DiseaseDescriptor':\n",
    "            value_type = 'Disease'\n",
    "            value_id_name = 'disease_id'\n",
    "            value_id = descriptor['value']['disease_id']\n",
    "        elif descr_type == 'GeneDescriptor':\n",
    "            value_type = 'Gene'\n",
    "            value_id_name = 'gene_id'\n",
    "            value_id = descriptor['value']['gene_id']\n",
    "        properties = \"\"\n",
    "        for key in ('id', 'label', 'description', 'xrefs', 'alternate_labels'):\n",
    "            value = descriptor.get(key)\n",
    "            if value:\n",
    "                properties += f'{key}:\"{value}\", '\n",
    "        if properties:\n",
    "            properties = properties[:-2]\n",
    "\n",
    "        try:\n",
    "            query = f'''\n",
    "            MERGE (descr:{descr_type} {{ {properties} }})\n",
    "            MERGE (value:{value_type} {{ {value_id_name}: \"{value_id}\" }})\n",
    "            MERGE (descr) -[:DESCRIBES]-> (value)\n",
    "            '''\n",
    "            tx.run(query)\n",
    "        except ServiceUnavailable as exception:\n",
    "            logging.error(f\"Failed to add Descriptor object\\nQuery: {query}\\n\"\n",
    "                          f\"Descriptor: {descriptor}\")\n",
    "            raise exception\n",
    "\n",
    "    @staticmethod\n",
    "    def _add_variant_descriptor(tx, descriptor: Dict):\n",
    "        \"\"\"Add variant descriptor object to DB.\n",
    "        TODO: evaluate using APOC functions\n",
    "        \"\"\"\n",
    "        value_type = 'Allele'\n",
    "        descriptor['value_state'] = json.dumps(descriptor['value']['state'])\n",
    "        descriptor['value_location'] = \\\n",
    "            json.dumps(descriptor['value']['location'])\n",
    "        descriptor['expressions'] = [json.dumps(e)\n",
    "                                     for e in descriptor['expressions']]\n",
    "        properties = \"\"\n",
    "        for key in ('id', 'label', 'description', 'xrefs', 'alternate_labels',\n",
    "                    'structural_type', \n",
    "                    #'expressions', \n",
    "                    'ref_allele_seq'):\n",
    "            value = descriptor.get(key)\n",
    "            if value:\n",
    "                properties += f'{key}:\"{value}\", '\n",
    "        if properties:\n",
    "            properties = properties[:-2]\n",
    "\n",
    "        query = f\"\"\"\n",
    "        MERGE (descr:VariationDescriptor\n",
    "            {{ {properties} }})\n",
    "        MERGE (value:{value_type}:Variation\n",
    "            {{state:$value_state,\n",
    "              location:$value_location}})\n",
    "        MERGE (gene_context:GeneDescriptor {{id:$gene_context}} )\n",
    "        MERGE (descr) -[:DESCRIBES]-> (value)\n",
    "        MERGE (descr) -[:HAS_GENE] -> (gene_context);\n",
    "        \"\"\"\n",
    "        try:\n",
    "            tx.run(query, **descriptor)\n",
    "        except ServiceUnavailable as exception:\n",
    "            logging.error(f\"Failed to add Variant Descriptor object\\nQuery: \"\n",
    "                          f\"{query}\\nDescriptor: {descriptor}\")\n",
    "            raise exception\n",
    "\n",
    "    @staticmethod\n",
    "    def _add_document(tx, document: Dict):\n",
    "        \"\"\"Add Document object to DB.\"\"\"\n",
    "        query = \"\"\"\n",
    "        MERGE (n:Document {id:$id, document_id:$document_id, label:$label,\n",
    "                           description:$description, xrefs:$xrefs});\n",
    "        \"\"\"\n",
    "        try:\n",
    "            tx.run(query, **document)\n",
    "        except ServiceUnavailable as exception:\n",
    "            logging.error(f\"Failed to add Document object\\n\"\n",
    "                          f\"Query: {query}\\nDocument: \"\n",
    "                          f\"{document}\")\n",
    "            raise exception\n",
    "\n",
    "    @staticmethod\n",
    "    def _add_evidence(tx, evidence: Dict):\n",
    "        \"\"\"Add evidence object to DB.\n",
    "\n",
    "        :param Dict evidence: Evidence object\n",
    "        \"\"\"\n",
    "        nonnull_values = [f'{key}:\"{evidence[key]}\"' for key \n",
    "                          in ('id', 'description', 'direction', 'evidence_level')\n",
    "                          if evidence[key]]\n",
    "        values_joined = ', '.join(nonnull_values)\n",
    "\n",
    "        query = f\"\"\"\n",
    "        MERGE (ev:Evidence {{ {values_joined} }})\n",
    "        MERGE (prop:Proposition {{_id:$proposition}})\n",
    "        MERGE (var:VariationDescriptor {{id:$variation_descriptor}})\n",
    "        MERGE (ther:TherapyDescriptor {{id:$therapy_descriptor}})\n",
    "        MERGE (dis:DiseaseDescriptor {{id:$disease_descriptor}})\n",
    "        MERGE (method:AssertionMethod {{id:$assertion_method}})\n",
    "        MERGE (doc:Document {{id:$document}})\n",
    "        MERGE (ev) -[:SUPPORTS]-> (prop)\n",
    "        MERGE (ev) -[:HAS_VARIANT]-> (var)\n",
    "        MERGE (ev) -[:HAS_THERAPY]-> (ther)\n",
    "        MERGE (ev) -[:HAS_DISEASE]-> (dis)\n",
    "        MERGE (ev) -[:USES_METHOD]-> (method)\n",
    "        MERGE (ev) -[:CITES]-> (doc)\n",
    "        \"\"\"\n",
    "        try:\n",
    "            tx.run(query, **evidence)\n",
    "        except ServiceUnavailable as exception:\n",
    "            logging.error(f\"Failed to add Evidence object\\n\"\n",
    "                          f\"Query: {query}\\nEvidence: {evidence}\")\n",
    "            raise exception\n",
    "\n",
    "g = Graph(uri=\"bolt://localhost:7687\", credentials=(\"neo4j\", \"admin\"))\n",
    "g.clear()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "from metakb import PROJECT_ROOT\n",
    "import json\n",
    "fpath = PROJECT_ROOT / 'analysis' / 'graph' / 'civic_cdm_updated.json'\n",
    "with open(fpath, 'r') as f:\n",
    "    items = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_blacklist = {'civic:gid30', 'civic:gid52'}\n",
    "therapies_blacklist = {'civic:tid237', 'civic:tid101', 'civic:tid326', 'civic:tid457', 'civic:tid14', 'civic:tid144', 'civic:tid138', 'civic:tid260', 'civic:tid505'}\n",
    "ev_blacklist = {'civic:eid8190', 'civic:eid2924', 'civic:eid2919', 'civic:eid8187', 'civic:eid8186', 'civic:eid2922', 'civic:eid8184', 'civic:eid2923', 'civic:eid8185', 'civic:eid8189', 'civic:eid1362'}\n",
    "count = 0\n",
    "for item in items:\n",
    "    if 'assertion' in item.keys():\n",
    "        continue\n",
    "    if any(filter(lambda x: x['id'] in genes_blacklist, item.get('gene_descriptors', []))):\n",
    "        continue\n",
    "    elif any(filter(lambda x: x['id'] in therapies_blacklist, item['therapy_descriptors'])):\n",
    "        continue\n",
    "    elif any(filter(lambda x: x['id'] in ev_blacklist, item['evidence'])):\n",
    "        continue\n",
    "    else:\n",
    "        g.add_transformed_data(item)\n",
    "        count += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1021"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "metakb",
   "language": "python",
   "name": "metakb"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
