{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Join and Normalize GENIE Data\n",
    "As part of the MetaKB v2 analysis, we will utilize part of the GENIE dataset to look at available interpretations in the aggregate knowledgebase for concepts as determined by real patient samples. We will join the mutation and clinical datasets, normalize values, and quantify interpretations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>...</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>Polyphen_Prediction</th>\n",
       "      <th>Polyphen_Score</th>\n",
       "      <th>SIFT_Prediction</th>\n",
       "      <th>SIFT_Score</th>\n",
       "      <th>SWISSPROT</th>\n",
       "      <th>n_depth</th>\n",
       "      <th>t_depth</th>\n",
       "      <th>Annotation_Status</th>\n",
       "      <th>mutationInCis_Flag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>12</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>0.991</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1623.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>673.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>7</td>\n",
       "      <td>140453136</td>\n",
       "      <td>140453136</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>0.963</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1031.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>1956.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>7</td>\n",
       "      <td>55249071</td>\n",
       "      <td>55249071</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>1.000</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>692.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7157.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>possibly_damaging</td>\n",
       "      <td>0.643</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>930.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>4893.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>115256529</td>\n",
       "      <td>115256529</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.251</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2277.0</td>\n",
       "      <td>SUCCESS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  Start_Position  \\\n",
       "0        KRAS          3845.0    JHU     GRCh37         12        25398285   \n",
       "1        BRAF           673.0    JHU     GRCh37          7       140453136   \n",
       "2        EGFR          1956.0    JHU     GRCh37          7        55249071   \n",
       "3        TP53          7157.0    JHU     GRCh37         17         7577120   \n",
       "4        NRAS          4893.0    JHU     GRCh37          1       115256529   \n",
       "\n",
       "   End_Position Strand       Consequence Variant_Classification  ... FILTER  \\\n",
       "0      25398285      +  missense_variant      Missense_Mutation  ...   PASS   \n",
       "1     140453136      +  missense_variant      Missense_Mutation  ...   PASS   \n",
       "2      55249071      +  missense_variant      Missense_Mutation  ...   PASS   \n",
       "3       7577120      +  missense_variant      Missense_Mutation  ...   PASS   \n",
       "4     115256529      +  missense_variant      Missense_Mutation  ...   PASS   \n",
       "\n",
       "  Polyphen_Prediction Polyphen_Score SIFT_Prediction SIFT_Score  SWISSPROT  \\\n",
       "0   probably_damaging          0.991     deleterious       0.04        NaN   \n",
       "1   probably_damaging          0.963     deleterious       0.00        NaN   \n",
       "2   probably_damaging          1.000     deleterious       0.00        NaN   \n",
       "3   possibly_damaging          0.643       tolerated       0.13        NaN   \n",
       "4              benign          0.251       tolerated       0.06        NaN   \n",
       "\n",
       "  n_depth t_depth Annotation_Status mutationInCis_Flag  \n",
       "0     NaN  1623.0           SUCCESS              False  \n",
       "1     NaN  1031.0           SUCCESS              False  \n",
       "2     NaN   692.0           SUCCESS              False  \n",
       "3     NaN   930.0           SUCCESS              False  \n",
       "4     NaN  2277.0           SUCCESS              False  \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Project GENIE Mutation Dataset\n",
    "muts = pd.read_csv('genie_data/data_mutations_extended.txt',sep='\\t', low_memory=False)\n",
    "muts[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>SAMPLE_ID</th>\n",
       "      <th>AGE_AT_SEQ_REPORT</th>\n",
       "      <th>ONCOTREE_CODE</th>\n",
       "      <th>SAMPLE_TYPE</th>\n",
       "      <th>SEQ_ASSAY_ID</th>\n",
       "      <th>CANCER_TYPE</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "      <th>SAMPLE_TYPE_DETAILED</th>\n",
       "      <th>SAMPLE_CLASS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GENIE-VICC-101416</td>\n",
       "      <td>GENIE-VICC-101416-unk-1</td>\n",
       "      <td>52</td>\n",
       "      <td>MAAP</td>\n",
       "      <td>Primary</td>\n",
       "      <td>VICC-01-T7</td>\n",
       "      <td>Appendiceal Cancer</td>\n",
       "      <td>Mucinous Adenocarcinoma of the Appendix</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GENIE-VICC-102225</td>\n",
       "      <td>GENIE-VICC-102225-unk-1</td>\n",
       "      <td>31</td>\n",
       "      <td>READ</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>VICC-01-T7</td>\n",
       "      <td>Colorectal Cancer</td>\n",
       "      <td>Rectal Adenocarcinoma</td>\n",
       "      <td>Metastasis site unspecified</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GENIE-VICC-102424</td>\n",
       "      <td>GENIE-VICC-102424-unk-1</td>\n",
       "      <td>63</td>\n",
       "      <td>SCCNOS</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>VICC-01-T7</td>\n",
       "      <td>Cancer of Unknown Primary</td>\n",
       "      <td>Squamous Cell Carcinoma, NOS</td>\n",
       "      <td>Metastasis site unspecified</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GENIE-VICC-102966</td>\n",
       "      <td>GENIE-VICC-102966-unk-1</td>\n",
       "      <td>64</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>VICC-01-T7</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Metastasis site unspecified</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GENIE-VICC-103244</td>\n",
       "      <td>GENIE-VICC-103244-unk-1</td>\n",
       "      <td>50</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Primary</td>\n",
       "      <td>VICC-01-T5A</td>\n",
       "      <td>Breast Cancer</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          PATIENT_ID                SAMPLE_ID AGE_AT_SEQ_REPORT ONCOTREE_CODE  \\\n",
       "0  GENIE-VICC-101416  GENIE-VICC-101416-unk-1                52          MAAP   \n",
       "1  GENIE-VICC-102225  GENIE-VICC-102225-unk-1                31          READ   \n",
       "2  GENIE-VICC-102424  GENIE-VICC-102424-unk-1                63        SCCNOS   \n",
       "3  GENIE-VICC-102966  GENIE-VICC-102966-unk-1                64          LUAD   \n",
       "4  GENIE-VICC-103244  GENIE-VICC-103244-unk-1                50           IDC   \n",
       "\n",
       "  SAMPLE_TYPE SEQ_ASSAY_ID                 CANCER_TYPE  \\\n",
       "0     Primary   VICC-01-T7          Appendiceal Cancer   \n",
       "1  Metastasis   VICC-01-T7           Colorectal Cancer   \n",
       "2  Metastasis   VICC-01-T7   Cancer of Unknown Primary   \n",
       "3  Metastasis   VICC-01-T7  Non-Small Cell Lung Cancer   \n",
       "4     Primary  VICC-01-T5A               Breast Cancer   \n",
       "\n",
       "                      CANCER_TYPE_DETAILED         SAMPLE_TYPE_DETAILED  \\\n",
       "0  Mucinous Adenocarcinoma of the Appendix                Primary tumor   \n",
       "1                    Rectal Adenocarcinoma  Metastasis site unspecified   \n",
       "2             Squamous Cell Carcinoma, NOS  Metastasis site unspecified   \n",
       "3                      Lung Adenocarcinoma  Metastasis site unspecified   \n",
       "4         Breast Invasive Ductal Carcinoma                Primary tumor   \n",
       "\n",
       "  SAMPLE_CLASS  \n",
       "0        Tumor  \n",
       "1        Tumor  \n",
       "2        Tumor  \n",
       "3        Tumor  \n",
       "4        Tumor  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Project GENIE Clinical Dataset\n",
    "clin = pd.read_csv('genie_data/data_clinical_sample.txt',sep='\\t', low_memory=False, header=4)\n",
    "clin[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>...</th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>SAMPLE_ID</th>\n",
       "      <th>AGE_AT_SEQ_REPORT</th>\n",
       "      <th>ONCOTREE_CODE</th>\n",
       "      <th>SAMPLE_TYPE</th>\n",
       "      <th>SEQ_ASSAY_ID</th>\n",
       "      <th>CANCER_TYPE</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "      <th>SAMPLE_TYPE_DETAILED</th>\n",
       "      <th>SAMPLE_CLASS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>12</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>GENIE-JHU-00006</td>\n",
       "      <td>GENIE-JHU-00006-00185</td>\n",
       "      <td>61</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>673.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>7</td>\n",
       "      <td>140453136</td>\n",
       "      <td>140453136</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>GENIE-JHU-00006</td>\n",
       "      <td>GENIE-JHU-00006-00185</td>\n",
       "      <td>61</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>1956.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>7</td>\n",
       "      <td>55249071</td>\n",
       "      <td>55249071</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>GENIE-JHU-00006</td>\n",
       "      <td>GENIE-JHU-00006-00185</td>\n",
       "      <td>61</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7157.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>GENIE-JHU-00006</td>\n",
       "      <td>GENIE-JHU-00006-00185</td>\n",
       "      <td>61</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>4893.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>115256529</td>\n",
       "      <td>115256529</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>GENIE-JHU-00006</td>\n",
       "      <td>GENIE-JHU-00006-00185</td>\n",
       "      <td>61</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 74 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  Start_Position  \\\n",
       "0        KRAS          3845.0    JHU     GRCh37         12        25398285   \n",
       "1        BRAF           673.0    JHU     GRCh37          7       140453136   \n",
       "2        EGFR          1956.0    JHU     GRCh37          7        55249071   \n",
       "3        TP53          7157.0    JHU     GRCh37         17         7577120   \n",
       "4        NRAS          4893.0    JHU     GRCh37          1       115256529   \n",
       "\n",
       "   End_Position Strand       Consequence Variant_Classification  ...  \\\n",
       "0      25398285      +  missense_variant      Missense_Mutation  ...   \n",
       "1     140453136      +  missense_variant      Missense_Mutation  ...   \n",
       "2      55249071      +  missense_variant      Missense_Mutation  ...   \n",
       "3       7577120      +  missense_variant      Missense_Mutation  ...   \n",
       "4     115256529      +  missense_variant      Missense_Mutation  ...   \n",
       "\n",
       "        PATIENT_ID              SAMPLE_ID AGE_AT_SEQ_REPORT ONCOTREE_CODE  \\\n",
       "0  GENIE-JHU-00006  GENIE-JHU-00006-00185                61          LUAD   \n",
       "1  GENIE-JHU-00006  GENIE-JHU-00006-00185                61          LUAD   \n",
       "2  GENIE-JHU-00006  GENIE-JHU-00006-00185                61          LUAD   \n",
       "3  GENIE-JHU-00006  GENIE-JHU-00006-00185                61          LUAD   \n",
       "4  GENIE-JHU-00006  GENIE-JHU-00006-00185                61          LUAD   \n",
       "\n",
       "  SAMPLE_TYPE  SEQ_ASSAY_ID                 CANCER_TYPE CANCER_TYPE_DETAILED  \\\n",
       "0     Primary      JHU-50GP  Non-Small Cell Lung Cancer  Lung Adenocarcinoma   \n",
       "1     Primary      JHU-50GP  Non-Small Cell Lung Cancer  Lung Adenocarcinoma   \n",
       "2     Primary      JHU-50GP  Non-Small Cell Lung Cancer  Lung Adenocarcinoma   \n",
       "3     Primary      JHU-50GP  Non-Small Cell Lung Cancer  Lung Adenocarcinoma   \n",
       "4     Primary      JHU-50GP  Non-Small Cell Lung Cancer  Lung Adenocarcinoma   \n",
       "\n",
       "  SAMPLE_TYPE_DETAILED SAMPLE_CLASS  \n",
       "0        Primary tumor        Tumor  \n",
       "1        Primary tumor        Tumor  \n",
       "2        Primary tumor        Tumor  \n",
       "3        Primary tumor        Tumor  \n",
       "4        Primary tumor        Tumor  \n",
       "\n",
       "[5 rows x 74 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Join the two data sets on patient identifiers\n",
    "df = pd.merge(muts, clin, left_on='Tumor_Sample_Barcode', right_on='SAMPLE_ID', how=\"left\")\n",
    "df[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'3_prime_UTR_variant',\n",
       " '5_prime_UTR_variant',\n",
       " 'NMD_transcript_variant',\n",
       " 'coding_sequence_variant',\n",
       " 'downstream_gene_variant',\n",
       " 'frameshift_variant',\n",
       " 'incomplete_terminal_codon_variant',\n",
       " 'inframe_deletion',\n",
       " 'inframe_insertion',\n",
       " 'intron_variant',\n",
       " 'mature_miRNA_variant',\n",
       " 'missense_variant',\n",
       " 'non_coding_transcript_exon_variant',\n",
       " 'non_coding_transcript_variant',\n",
       " 'protein_altering_variant',\n",
       " 'splice_acceptor_variant',\n",
       " 'splice_donor_variant',\n",
       " 'splice_region_variant',\n",
       " 'start_lost',\n",
       " 'start_retained_variant',\n",
       " 'stop_gained',\n",
       " 'stop_lost',\n",
       " 'stop_retained_variant',\n",
       " 'synonymous_variant',\n",
       " 'upstream_gene_variant'}"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# From MetaKB v1 paper --> Exclude silent, 3' Flank, 3'UTR, 5'Flank, 5'UR, Intron, Splice_region\n",
    "# 3_prime_UTR_variant, 5_prime_UTR_variant, splice_region_variant, splice_acceptor_variant, splice_donor_variant, intron_variant, synonymous_variant \n",
    "all_consequences = []\n",
    "for entry in df['Consequence']:\n",
    "    if isinstance(entry,float):\n",
    "        pass\n",
    "    else:\n",
    "        consequences = entry.split(',')\n",
    "        all_consequences.extend(consequences)\n",
    "set(all_consequences) \n",
    "\n",
    "# TODO: how do start/stop_retained variants function? are these silent?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>...</th>\n",
       "      <th>SAMPLE_ID</th>\n",
       "      <th>AGE_AT_SEQ_REPORT</th>\n",
       "      <th>ONCOTREE_CODE</th>\n",
       "      <th>SAMPLE_TYPE</th>\n",
       "      <th>SEQ_ASSAY_ID</th>\n",
       "      <th>CANCER_TYPE</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "      <th>SAMPLE_TYPE_DETAILED</th>\n",
       "      <th>SAMPLE_CLASS</th>\n",
       "      <th>Exclusion?</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>12</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>GENIE-JHU-00006-00185</td>\n",
       "      <td>61</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>673.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>7</td>\n",
       "      <td>140453136</td>\n",
       "      <td>140453136</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>GENIE-JHU-00006-00185</td>\n",
       "      <td>61</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>1956.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>7</td>\n",
       "      <td>55249071</td>\n",
       "      <td>55249071</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>GENIE-JHU-00006-00185</td>\n",
       "      <td>61</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7157.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>GENIE-JHU-00006-00185</td>\n",
       "      <td>61</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>4893.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>115256529</td>\n",
       "      <td>115256529</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>GENIE-JHU-00006-00185</td>\n",
       "      <td>61</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 75 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  Start_Position  \\\n",
       "0        KRAS          3845.0    JHU     GRCh37         12        25398285   \n",
       "1        BRAF           673.0    JHU     GRCh37          7       140453136   \n",
       "2        EGFR          1956.0    JHU     GRCh37          7        55249071   \n",
       "3        TP53          7157.0    JHU     GRCh37         17         7577120   \n",
       "4        NRAS          4893.0    JHU     GRCh37          1       115256529   \n",
       "\n",
       "   End_Position Strand       Consequence Variant_Classification  ...  \\\n",
       "0      25398285      +  missense_variant      Missense_Mutation  ...   \n",
       "1     140453136      +  missense_variant      Missense_Mutation  ...   \n",
       "2      55249071      +  missense_variant      Missense_Mutation  ...   \n",
       "3       7577120      +  missense_variant      Missense_Mutation  ...   \n",
       "4     115256529      +  missense_variant      Missense_Mutation  ...   \n",
       "\n",
       "               SAMPLE_ID AGE_AT_SEQ_REPORT ONCOTREE_CODE SAMPLE_TYPE  \\\n",
       "0  GENIE-JHU-00006-00185                61          LUAD     Primary   \n",
       "1  GENIE-JHU-00006-00185                61          LUAD     Primary   \n",
       "2  GENIE-JHU-00006-00185                61          LUAD     Primary   \n",
       "3  GENIE-JHU-00006-00185                61          LUAD     Primary   \n",
       "4  GENIE-JHU-00006-00185                61          LUAD     Primary   \n",
       "\n",
       "  SEQ_ASSAY_ID                 CANCER_TYPE CANCER_TYPE_DETAILED  \\\n",
       "0     JHU-50GP  Non-Small Cell Lung Cancer  Lung Adenocarcinoma   \n",
       "1     JHU-50GP  Non-Small Cell Lung Cancer  Lung Adenocarcinoma   \n",
       "2     JHU-50GP  Non-Small Cell Lung Cancer  Lung Adenocarcinoma   \n",
       "3     JHU-50GP  Non-Small Cell Lung Cancer  Lung Adenocarcinoma   \n",
       "4     JHU-50GP  Non-Small Cell Lung Cancer  Lung Adenocarcinoma   \n",
       "\n",
       "  SAMPLE_TYPE_DETAILED SAMPLE_CLASS Exclusion?  \n",
       "0        Primary tumor        Tumor      False  \n",
       "1        Primary tumor        Tumor      False  \n",
       "2        Primary tumor        Tumor      False  \n",
       "3        Primary tumor        Tumor      False  \n",
       "4        Primary tumor        Tumor      False  \n",
       "\n",
       "[5 rows x 75 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No Consequence (NA) also means silent? Drop!\n",
    "df = df.dropna(subset=['Consequence']).reset_index(drop=True)\n",
    "df[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          False\n",
       "1          False\n",
       "2          False\n",
       "3          False\n",
       "4          False\n",
       "           ...  \n",
       "2264728     True\n",
       "2264729    False\n",
       "2264730    False\n",
       "2264731    False\n",
       "2264732    False\n",
       "Name: Exclusion?, Length: 2264733, dtype: bool"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Using the above lists, excluse silent consequences \n",
    "exclusion_criteria = ['3_prime_UTR_variant', \n",
    "                     '5_prime_UTR_variant', \n",
    "                     'splice_region_variant', \n",
    "                     'splice_acceptor_variant', \n",
    "                     'splice_donor_variant', \n",
    "                     'intron_variant', \n",
    "                     'synonymous_variant']\n",
    "\n",
    "df['Exclusion?'] = False\n",
    "for idx, row in df.iterrows():\n",
    "    consequences = row['Consequence'].split(',')\n",
    "    for consequence in consequences:\n",
    "        if consequence in exclusion_criteria:\n",
    "            df.at[idx, 'Exclusion?'] = True\n",
    "\n",
    "df = df[df['Exclusion?']==False].reset_index(drop=True)\n",
    "df[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Hugo_Symbol', 'Entrez_Gene_Id', 'Center', 'NCBI_Build', 'Chromosome',\n",
       "       'Start_Position', 'End_Position', 'Strand', 'Consequence',\n",
       "       'Variant_Classification', 'Variant_Type', 'Reference_Allele',\n",
       "       'Tumor_Seq_Allele1', 'Tumor_Seq_Allele2', 'dbSNP_RS',\n",
       "       'dbSNP_Val_Status', 'Tumor_Sample_Barcode',\n",
       "       'Matched_Norm_Sample_Barcode', 'Match_Norm_Seq_Allele1',\n",
       "       'Match_Norm_Seq_Allele2', 'Tumor_Validation_Allele1',\n",
       "       'Tumor_Validation_Allele2', 'Match_Norm_Validation_Allele1',\n",
       "       'Match_Norm_Validation_Allele2', 'Verification_Status',\n",
       "       'Validation_Status', 'Mutation_Status', 'Sequencing_Phase',\n",
       "       'Sequence_Source', 'Validation_Method', 'Score', 'BAM_File',\n",
       "       'Sequencer', 't_ref_count', 't_alt_count', 'n_ref_count', 'n_alt_count',\n",
       "       'HGVSc', 'HGVSp', 'HGVSp_Short', 'Transcript_ID', 'RefSeq',\n",
       "       'Protein_position', 'Codons', 'Exon_Number', 'gnomAD_AF',\n",
       "       'gnomAD_AFR_AF', 'gnomAD_AMR_AF', 'gnomAD_ASJ_AF', 'gnomAD_EAS_AF',\n",
       "       'gnomAD_FIN_AF', 'gnomAD_NFE_AF', 'gnomAD_OTH_AF', 'gnomAD_SAS_AF',\n",
       "       'FILTER', 'Polyphen_Prediction', 'Polyphen_Score', 'SIFT_Prediction',\n",
       "       'SIFT_Score', 'SWISSPROT', 'n_depth', 't_depth', 'Annotation_Status',\n",
       "       'mutationInCis_Flag', 'PATIENT_ID', 'SAMPLE_ID', 'AGE_AT_SEQ_REPORT',\n",
       "       'ONCOTREE_CODE', 'SAMPLE_TYPE', 'SEQ_ASSAY_ID', 'CANCER_TYPE',\n",
       "       'CANCER_TYPE_DETAILED', 'SAMPLE_TYPE_DETAILED', 'SAMPLE_CLASS',\n",
       "       'Exclusion?'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get concept IDs for everything attached to the patients by the following corresponding fields\n",
    "# Variant -> HGVSc\n",
    "# Gene -> Hugo_Symbol\n",
    "# Disease -> CANCER_TYPE\n",
    "df.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Normalize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>...</th>\n",
       "      <th>AGE_AT_SEQ_REPORT</th>\n",
       "      <th>ONCOTREE_CODE</th>\n",
       "      <th>SAMPLE_TYPE</th>\n",
       "      <th>SEQ_ASSAY_ID</th>\n",
       "      <th>CANCER_TYPE</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "      <th>SAMPLE_TYPE_DETAILED</th>\n",
       "      <th>SAMPLE_CLASS</th>\n",
       "      <th>Exclusion?</th>\n",
       "      <th>disease_concept_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>12</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>61</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>673.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>7</td>\n",
       "      <td>140453136</td>\n",
       "      <td>140453136</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>61</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>1956.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>7</td>\n",
       "      <td>55249071</td>\n",
       "      <td>55249071</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>61</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7157.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>61</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>4893.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>115256529</td>\n",
       "      <td>115256529</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>61</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 76 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  Start_Position  \\\n",
       "0        KRAS          3845.0    JHU     GRCh37         12        25398285   \n",
       "1        BRAF           673.0    JHU     GRCh37          7       140453136   \n",
       "2        EGFR          1956.0    JHU     GRCh37          7        55249071   \n",
       "3        TP53          7157.0    JHU     GRCh37         17         7577120   \n",
       "4        NRAS          4893.0    JHU     GRCh37          1       115256529   \n",
       "\n",
       "   End_Position Strand       Consequence Variant_Classification  ...  \\\n",
       "0      25398285      +  missense_variant      Missense_Mutation  ...   \n",
       "1     140453136      +  missense_variant      Missense_Mutation  ...   \n",
       "2      55249071      +  missense_variant      Missense_Mutation  ...   \n",
       "3       7577120      +  missense_variant      Missense_Mutation  ...   \n",
       "4     115256529      +  missense_variant      Missense_Mutation  ...   \n",
       "\n",
       "  AGE_AT_SEQ_REPORT ONCOTREE_CODE SAMPLE_TYPE SEQ_ASSAY_ID  \\\n",
       "0                61          LUAD     Primary     JHU-50GP   \n",
       "1                61          LUAD     Primary     JHU-50GP   \n",
       "2                61          LUAD     Primary     JHU-50GP   \n",
       "3                61          LUAD     Primary     JHU-50GP   \n",
       "4                61          LUAD     Primary     JHU-50GP   \n",
       "\n",
       "                  CANCER_TYPE  CANCER_TYPE_DETAILED SAMPLE_TYPE_DETAILED  \\\n",
       "0  Non-Small Cell Lung Cancer   Lung Adenocarcinoma        Primary tumor   \n",
       "1  Non-Small Cell Lung Cancer   Lung Adenocarcinoma        Primary tumor   \n",
       "2  Non-Small Cell Lung Cancer   Lung Adenocarcinoma        Primary tumor   \n",
       "3  Non-Small Cell Lung Cancer   Lung Adenocarcinoma        Primary tumor   \n",
       "4  Non-Small Cell Lung Cancer   Lung Adenocarcinoma        Primary tumor   \n",
       "\n",
       "  SAMPLE_CLASS Exclusion?            disease_concept_id  \n",
       "0        Tumor      False  normalize.disease.ncit:C2926  \n",
       "1        Tumor      False  normalize.disease.ncit:C2926  \n",
       "2        Tumor      False  normalize.disease.ncit:C2926  \n",
       "3        Tumor      False  normalize.disease.ncit:C2926  \n",
       "4        Tumor      False  normalize.disease.ncit:C2926  \n",
       "\n",
       "[5 rows x 76 columns]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import requests \n",
    "\n",
    "def normalize_disease(term):\n",
    "    r = requests.get(f'https://normalize.cancervariants.org/disease/normalize?q={term}')\n",
    "    # print(r.json())\n",
    "    if r.status_code == 200:\n",
    "        try:\n",
    "            identifier = r.json()['disease']['id']\n",
    "        except:\n",
    "            identifier = None\n",
    "    return identifier\n",
    "\n",
    "unique_diseases = list(df['CANCER_TYPE'].unique())\n",
    "disease_identifiers = [None] * len(unique_diseases)\n",
    "disease_wordlist = pd.DataFrame({'unique_disease': unique_diseases,\n",
    "                                 'disease_identifier': disease_identifiers})\n",
    "disease_wordlist['disease_identifier'] = disease_wordlist['unique_disease'].apply(normalize_disease) \n",
    "disease_wordlist_map = disease_wordlist.set_index('unique_disease')['disease_identifier'] #TODO: why two variables?\n",
    "df['disease_concept_id'] = df['CANCER_TYPE'].map(disease_wordlist_map)\n",
    "df[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>...</th>\n",
       "      <th>ONCOTREE_CODE</th>\n",
       "      <th>SAMPLE_TYPE</th>\n",
       "      <th>SEQ_ASSAY_ID</th>\n",
       "      <th>CANCER_TYPE</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "      <th>SAMPLE_TYPE_DETAILED</th>\n",
       "      <th>SAMPLE_CLASS</th>\n",
       "      <th>Exclusion?</th>\n",
       "      <th>disease_concept_id</th>\n",
       "      <th>gene_concept_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>12</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "      <td>normalize.gene.hgnc:6407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>673.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>7</td>\n",
       "      <td>140453136</td>\n",
       "      <td>140453136</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "      <td>normalize.gene.hgnc:1097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>1956.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>7</td>\n",
       "      <td>55249071</td>\n",
       "      <td>55249071</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "      <td>normalize.gene.hgnc:3236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7157.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "      <td>normalize.gene.hgnc:11998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>4893.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>115256529</td>\n",
       "      <td>115256529</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "      <td>normalize.gene.hgnc:7989</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 77 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  Start_Position  \\\n",
       "0        KRAS          3845.0    JHU     GRCh37         12        25398285   \n",
       "1        BRAF           673.0    JHU     GRCh37          7       140453136   \n",
       "2        EGFR          1956.0    JHU     GRCh37          7        55249071   \n",
       "3        TP53          7157.0    JHU     GRCh37         17         7577120   \n",
       "4        NRAS          4893.0    JHU     GRCh37          1       115256529   \n",
       "\n",
       "   End_Position Strand       Consequence Variant_Classification  ...  \\\n",
       "0      25398285      +  missense_variant      Missense_Mutation  ...   \n",
       "1     140453136      +  missense_variant      Missense_Mutation  ...   \n",
       "2      55249071      +  missense_variant      Missense_Mutation  ...   \n",
       "3       7577120      +  missense_variant      Missense_Mutation  ...   \n",
       "4     115256529      +  missense_variant      Missense_Mutation  ...   \n",
       "\n",
       "  ONCOTREE_CODE SAMPLE_TYPE SEQ_ASSAY_ID                 CANCER_TYPE  \\\n",
       "0          LUAD     Primary     JHU-50GP  Non-Small Cell Lung Cancer   \n",
       "1          LUAD     Primary     JHU-50GP  Non-Small Cell Lung Cancer   \n",
       "2          LUAD     Primary     JHU-50GP  Non-Small Cell Lung Cancer   \n",
       "3          LUAD     Primary     JHU-50GP  Non-Small Cell Lung Cancer   \n",
       "4          LUAD     Primary     JHU-50GP  Non-Small Cell Lung Cancer   \n",
       "\n",
       "  CANCER_TYPE_DETAILED  SAMPLE_TYPE_DETAILED SAMPLE_CLASS Exclusion?  \\\n",
       "0  Lung Adenocarcinoma         Primary tumor        Tumor      False   \n",
       "1  Lung Adenocarcinoma         Primary tumor        Tumor      False   \n",
       "2  Lung Adenocarcinoma         Primary tumor        Tumor      False   \n",
       "3  Lung Adenocarcinoma         Primary tumor        Tumor      False   \n",
       "4  Lung Adenocarcinoma         Primary tumor        Tumor      False   \n",
       "\n",
       "             disease_concept_id            gene_concept_id  \n",
       "0  normalize.disease.ncit:C2926   normalize.gene.hgnc:6407  \n",
       "1  normalize.disease.ncit:C2926   normalize.gene.hgnc:1097  \n",
       "2  normalize.disease.ncit:C2926   normalize.gene.hgnc:3236  \n",
       "3  normalize.disease.ncit:C2926  normalize.gene.hgnc:11998  \n",
       "4  normalize.disease.ncit:C2926   normalize.gene.hgnc:7989  \n",
       "\n",
       "[5 rows x 77 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def normalize_gene(term):\n",
    "    r = requests.get(f'https://normalize.cancervariants.org/gene/normalize?q={term}')\n",
    "    # print(r.json())\n",
    "    if r.status_code == 200:\n",
    "        try:\n",
    "            identifier = r.json()['gene']['id']\n",
    "        except:\n",
    "            identifier = None\n",
    "    return identifier\n",
    "\n",
    "unique_genes = list(df['Hugo_Symbol'].unique())\n",
    "gene_identifiers = [None] * len(unique_genes)\n",
    "gene_wordlist = pd.DataFrame({'unique_gene': unique_genes,\n",
    "                                 'gene_identifier': gene_identifiers})\n",
    "\n",
    "gene_wordlist['gene_identifier'] = gene_wordlist['unique_gene'].apply(normalize_gene) \n",
    "gene_wordlist = gene_wordlist.set_index('unique_gene')['gene_identifier'] \n",
    "df['gene_concept_id'] = df['Hugo_Symbol'].map(gene_wordlist)\n",
    "df[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>...</th>\n",
       "      <th>SAMPLE_TYPE</th>\n",
       "      <th>SEQ_ASSAY_ID</th>\n",
       "      <th>CANCER_TYPE</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "      <th>SAMPLE_TYPE_DETAILED</th>\n",
       "      <th>SAMPLE_CLASS</th>\n",
       "      <th>Exclusion?</th>\n",
       "      <th>disease_concept_id</th>\n",
       "      <th>gene_concept_id</th>\n",
       "      <th>var_norm_query</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>12</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "      <td>normalize.gene.hgnc:6407</td>\n",
       "      <td>KRAS G12C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>673.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>7</td>\n",
       "      <td>140453136</td>\n",
       "      <td>140453136</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "      <td>normalize.gene.hgnc:1097</td>\n",
       "      <td>BRAF V600E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>1956.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>7</td>\n",
       "      <td>55249071</td>\n",
       "      <td>55249071</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "      <td>normalize.gene.hgnc:3236</td>\n",
       "      <td>EGFR T790M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7157.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "      <td>normalize.gene.hgnc:11998</td>\n",
       "      <td>TP53 R273H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>4893.0</td>\n",
       "      <td>JHU</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>115256529</td>\n",
       "      <td>115256529</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>Primary</td>\n",
       "      <td>JHU-50GP</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>False</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "      <td>normalize.gene.hgnc:7989</td>\n",
       "      <td>NRAS Q61R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1876019</th>\n",
       "      <td>TBX3</td>\n",
       "      <td>6926.0</td>\n",
       "      <td>PROV</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>12</td>\n",
       "      <td>115112292</td>\n",
       "      <td>115112292</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable or Heme</td>\n",
       "      <td>PROV-FOUNDATIONONELIQUIDCDX</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>Not applicable or hematologic malignancy</td>\n",
       "      <td>cfDNA</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>normalize.gene.hgnc:11602</td>\n",
       "      <td>TBX3 T483M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1876020</th>\n",
       "      <td>AR</td>\n",
       "      <td>367.0</td>\n",
       "      <td>PROV</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>X</td>\n",
       "      <td>66765014</td>\n",
       "      <td>66765014</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable or Heme</td>\n",
       "      <td>PROV-FOUNDATIONONELIQUIDCDX</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>Not applicable or hematologic malignancy</td>\n",
       "      <td>cfDNA</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>normalize.gene.hgnc:644</td>\n",
       "      <td>AR R9K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1876021</th>\n",
       "      <td>MSH3</td>\n",
       "      <td>4437.0</td>\n",
       "      <td>PROV</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>5</td>\n",
       "      <td>80021325</td>\n",
       "      <td>80021325</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable or Heme</td>\n",
       "      <td>PROV-FOUNDATIONONELIQUIDCDX</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>Not applicable or hematologic malignancy</td>\n",
       "      <td>cfDNA</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>normalize.gene.hgnc:7326</td>\n",
       "      <td>MSH3 Y465C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1876022</th>\n",
       "      <td>ATM</td>\n",
       "      <td>472.0</td>\n",
       "      <td>PROV</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>11</td>\n",
       "      <td>108098589</td>\n",
       "      <td>108098589</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable or Heme</td>\n",
       "      <td>PROV-FOUNDATIONONELIQUIDCDX</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>Not applicable or hematologic malignancy</td>\n",
       "      <td>cfDNA</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>normalize.gene.hgnc:795</td>\n",
       "      <td>ATM K53N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1876023</th>\n",
       "      <td>AR</td>\n",
       "      <td>367.0</td>\n",
       "      <td>PROV</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>X</td>\n",
       "      <td>66931295</td>\n",
       "      <td>66931295</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable or Heme</td>\n",
       "      <td>PROV-FOUNDATIONONELIQUIDCDX</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>Not applicable or hematologic malignancy</td>\n",
       "      <td>cfDNA</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>normalize.gene.hgnc:644</td>\n",
       "      <td>AR A646D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1876024 rows × 78 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  \\\n",
       "0              KRAS          3845.0    JHU     GRCh37         12   \n",
       "1              BRAF           673.0    JHU     GRCh37          7   \n",
       "2              EGFR          1956.0    JHU     GRCh37          7   \n",
       "3              TP53          7157.0    JHU     GRCh37         17   \n",
       "4              NRAS          4893.0    JHU     GRCh37          1   \n",
       "...             ...             ...    ...        ...        ...   \n",
       "1876019        TBX3          6926.0   PROV     GRCh37         12   \n",
       "1876020          AR           367.0   PROV     GRCh37          X   \n",
       "1876021        MSH3          4437.0   PROV     GRCh37          5   \n",
       "1876022         ATM           472.0   PROV     GRCh37         11   \n",
       "1876023          AR           367.0   PROV     GRCh37          X   \n",
       "\n",
       "         Start_Position  End_Position Strand       Consequence  \\\n",
       "0              25398285      25398285      +  missense_variant   \n",
       "1             140453136     140453136      +  missense_variant   \n",
       "2              55249071      55249071      +  missense_variant   \n",
       "3               7577120       7577120      +  missense_variant   \n",
       "4             115256529     115256529      +  missense_variant   \n",
       "...                 ...           ...    ...               ...   \n",
       "1876019       115112292     115112292      +  missense_variant   \n",
       "1876020        66765014      66765014      +  missense_variant   \n",
       "1876021        80021325      80021325      +  missense_variant   \n",
       "1876022       108098589     108098589      +  missense_variant   \n",
       "1876023        66931295      66931295      +  missense_variant   \n",
       "\n",
       "        Variant_Classification  ...             SAMPLE_TYPE  \\\n",
       "0            Missense_Mutation  ...                 Primary   \n",
       "1            Missense_Mutation  ...                 Primary   \n",
       "2            Missense_Mutation  ...                 Primary   \n",
       "3            Missense_Mutation  ...                 Primary   \n",
       "4            Missense_Mutation  ...                 Primary   \n",
       "...                        ...  ...                     ...   \n",
       "1876019      Missense_Mutation  ...  Not Applicable or Heme   \n",
       "1876020      Missense_Mutation  ...  Not Applicable or Heme   \n",
       "1876021      Missense_Mutation  ...  Not Applicable or Heme   \n",
       "1876022      Missense_Mutation  ...  Not Applicable or Heme   \n",
       "1876023      Missense_Mutation  ...  Not Applicable or Heme   \n",
       "\n",
       "                        SEQ_ASSAY_ID                 CANCER_TYPE  \\\n",
       "0                           JHU-50GP  Non-Small Cell Lung Cancer   \n",
       "1                           JHU-50GP  Non-Small Cell Lung Cancer   \n",
       "2                           JHU-50GP  Non-Small Cell Lung Cancer   \n",
       "3                           JHU-50GP  Non-Small Cell Lung Cancer   \n",
       "4                           JHU-50GP  Non-Small Cell Lung Cancer   \n",
       "...                              ...                         ...   \n",
       "1876019  PROV-FOUNDATIONONELIQUIDCDX                     UNKNOWN   \n",
       "1876020  PROV-FOUNDATIONONELIQUIDCDX                     UNKNOWN   \n",
       "1876021  PROV-FOUNDATIONONELIQUIDCDX                     UNKNOWN   \n",
       "1876022  PROV-FOUNDATIONONELIQUIDCDX                     UNKNOWN   \n",
       "1876023  PROV-FOUNDATIONONELIQUIDCDX                     UNKNOWN   \n",
       "\n",
       "        CANCER_TYPE_DETAILED                      SAMPLE_TYPE_DETAILED  \\\n",
       "0        Lung Adenocarcinoma                             Primary tumor   \n",
       "1        Lung Adenocarcinoma                             Primary tumor   \n",
       "2        Lung Adenocarcinoma                             Primary tumor   \n",
       "3        Lung Adenocarcinoma                             Primary tumor   \n",
       "4        Lung Adenocarcinoma                             Primary tumor   \n",
       "...                      ...                                       ...   \n",
       "1876019              UNKNOWN  Not applicable or hematologic malignancy   \n",
       "1876020              UNKNOWN  Not applicable or hematologic malignancy   \n",
       "1876021              UNKNOWN  Not applicable or hematologic malignancy   \n",
       "1876022              UNKNOWN  Not applicable or hematologic malignancy   \n",
       "1876023              UNKNOWN  Not applicable or hematologic malignancy   \n",
       "\n",
       "         SAMPLE_CLASS Exclusion?            disease_concept_id  \\\n",
       "0               Tumor      False  normalize.disease.ncit:C2926   \n",
       "1               Tumor      False  normalize.disease.ncit:C2926   \n",
       "2               Tumor      False  normalize.disease.ncit:C2926   \n",
       "3               Tumor      False  normalize.disease.ncit:C2926   \n",
       "4               Tumor      False  normalize.disease.ncit:C2926   \n",
       "...               ...        ...                           ...   \n",
       "1876019         cfDNA      False                          None   \n",
       "1876020         cfDNA      False                          None   \n",
       "1876021         cfDNA      False                          None   \n",
       "1876022         cfDNA      False                          None   \n",
       "1876023         cfDNA      False                          None   \n",
       "\n",
       "                   gene_concept_id var_norm_query  \n",
       "0         normalize.gene.hgnc:6407      KRAS G12C  \n",
       "1         normalize.gene.hgnc:1097     BRAF V600E  \n",
       "2         normalize.gene.hgnc:3236     EGFR T790M  \n",
       "3        normalize.gene.hgnc:11998     TP53 R273H  \n",
       "4         normalize.gene.hgnc:7989      NRAS Q61R  \n",
       "...                            ...            ...  \n",
       "1876019  normalize.gene.hgnc:11602     TBX3 T483M  \n",
       "1876020    normalize.gene.hgnc:644         AR R9K  \n",
       "1876021   normalize.gene.hgnc:7326     MSH3 Y465C  \n",
       "1876022    normalize.gene.hgnc:795       ATM K53N  \n",
       "1876023    normalize.gene.hgnc:644       AR A646D  \n",
       "\n",
       "[1876024 rows x 78 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def build_varnorm_query(HGVSp, Hugo_Symbol):\n",
    "    if isinstance(HGVSp, float):\n",
    "        return None\n",
    "    if isinstance(Hugo_Symbol, float):\n",
    "        return None\n",
    "    return f'{Hugo_Symbol} {HGVSp.split(\".\")[1]}'\n",
    "    \n",
    "df['var_norm_query'] = df.apply(lambda x: build_varnorm_query(x['HGVSp_Short'], x['Hugo_Symbol']), axis=1)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "767237"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(list(df['var_norm_query'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Get the variation normalization service up and running locally, 767237 unique values is too many to run on the Fast API\n",
    "# Kori is updating some README instructions for loading Variation Normalizer locally\n",
    "# In the meantime, just use the var_norm_query field to build out the notebook.\n",
    "\n",
    "# from tqdm import tqdm\n",
    "# tqdm.pandas()\n",
    "# def normalize_variation(term):\n",
    "#     r = requests.get(f'https://normalize.cancervariants.org/variation/normalize?q={term}')\n",
    "#     # print(r.json())\n",
    "#     if r.status_code == 200:\n",
    "#         try:\n",
    "#             identifier = r.json()['variation']['id']\n",
    "#         except:\n",
    "#             identifier = None\n",
    "#     return identifier\n",
    "\n",
    "# unique_variations = list(df['var_norm_query'].unique())\n",
    "# variation_identifiers = [None] * len(unique_variations)\n",
    "# variation_wordlist = pd.DataFrame({'unique_variation': unique_variations,\n",
    "#                                  'variation_identifier': variation_identifiers})\n",
    "\n",
    "# variation_wordlist['variation_identifier'] = variation_wordlist['unique_variation'].progress_apply(normalize_variation) \n",
    "# variation_wordlist = variation_wordlist.set_index('unique_variation')['variation_identifier'] \n",
    "# df['variation_concept_id'] = df['var_norm_query'].map(gene_wordlist)\n",
    "# df[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>disease_concept_id</th>\n",
       "      <th>gene_concept_id</th>\n",
       "      <th>var_norm_query</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GENIE-JHU-00006</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "      <td>normalize.gene.hgnc:6407</td>\n",
       "      <td>KRAS G12C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GENIE-JHU-00006</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "      <td>normalize.gene.hgnc:1097</td>\n",
       "      <td>BRAF V600E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GENIE-JHU-00006</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "      <td>normalize.gene.hgnc:3236</td>\n",
       "      <td>EGFR T790M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GENIE-JHU-00006</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "      <td>normalize.gene.hgnc:11998</td>\n",
       "      <td>TP53 R273H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GENIE-JHU-00006</td>\n",
       "      <td>normalize.disease.ncit:C2926</td>\n",
       "      <td>normalize.gene.hgnc:7989</td>\n",
       "      <td>NRAS Q61R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1876019</th>\n",
       "      <td>GENIE-PROV-9a7b04d343</td>\n",
       "      <td>None</td>\n",
       "      <td>normalize.gene.hgnc:11602</td>\n",
       "      <td>TBX3 T483M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1876020</th>\n",
       "      <td>GENIE-PROV-9a7b04d343</td>\n",
       "      <td>None</td>\n",
       "      <td>normalize.gene.hgnc:644</td>\n",
       "      <td>AR R9K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1876021</th>\n",
       "      <td>GENIE-PROV-9a7b04d343</td>\n",
       "      <td>None</td>\n",
       "      <td>normalize.gene.hgnc:7326</td>\n",
       "      <td>MSH3 Y465C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1876022</th>\n",
       "      <td>GENIE-PROV-9a7b04d343</td>\n",
       "      <td>None</td>\n",
       "      <td>normalize.gene.hgnc:795</td>\n",
       "      <td>ATM K53N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1876023</th>\n",
       "      <td>GENIE-PROV-9a7b04d343</td>\n",
       "      <td>None</td>\n",
       "      <td>normalize.gene.hgnc:644</td>\n",
       "      <td>AR A646D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1876024 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    PATIENT_ID            disease_concept_id  \\\n",
       "0              GENIE-JHU-00006  normalize.disease.ncit:C2926   \n",
       "1              GENIE-JHU-00006  normalize.disease.ncit:C2926   \n",
       "2              GENIE-JHU-00006  normalize.disease.ncit:C2926   \n",
       "3              GENIE-JHU-00006  normalize.disease.ncit:C2926   \n",
       "4              GENIE-JHU-00006  normalize.disease.ncit:C2926   \n",
       "...                        ...                           ...   \n",
       "1876019  GENIE-PROV-9a7b04d343                          None   \n",
       "1876020  GENIE-PROV-9a7b04d343                          None   \n",
       "1876021  GENIE-PROV-9a7b04d343                          None   \n",
       "1876022  GENIE-PROV-9a7b04d343                          None   \n",
       "1876023  GENIE-PROV-9a7b04d343                          None   \n",
       "\n",
       "                   gene_concept_id var_norm_query  \n",
       "0         normalize.gene.hgnc:6407      KRAS G12C  \n",
       "1         normalize.gene.hgnc:1097     BRAF V600E  \n",
       "2         normalize.gene.hgnc:3236     EGFR T790M  \n",
       "3        normalize.gene.hgnc:11998     TP53 R273H  \n",
       "4         normalize.gene.hgnc:7989      NRAS Q61R  \n",
       "...                            ...            ...  \n",
       "1876019  normalize.gene.hgnc:11602     TBX3 T483M  \n",
       "1876020    normalize.gene.hgnc:644         AR R9K  \n",
       "1876021   normalize.gene.hgnc:7326     MSH3 Y465C  \n",
       "1876022    normalize.gene.hgnc:795       ATM K53N  \n",
       "1876023    normalize.gene.hgnc:644       AR A646D  \n",
       "\n",
       "[1876024 rows x 4 columns]"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# These are our relevant fields for quantification\n",
    "df[['PATIENT_ID','disease_concept_id','gene_concept_id','var_norm_query']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the MetaKB v2 Dataset\n",
    "The GENIE dataset is normalized and ready for comparison, now load in the MetaKB v2 dataset to quantify available interpretations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "3.12.2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
