{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Investigate MetaKB Dataset: Therapeutics & Combination Therapies\n",
    "In order to better understand the MetaKB aggregate dataset, perform graph-directed lookups via neo4j to identify potential insights or avenues of discussion that might not previously be accessible.\n",
    "  \n",
    "This notebook focuses on exploring the Therapeutics & Combination Therapies datasets across different variants, studies, and gene contexts. Initial ideas are to look at FDA approval for therapies, diseases in use, and off-label use cases on a variant to variant or gene to gene basis.\n",
    "  \n",
    "**Current Data Version**: 5.20.0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Grab Therapeutics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from neo4j import GraphDatabase\n",
    "\n",
    "# Function to create a connection to the Neo4j database\n",
    "def create_db_connection(uri, user, password):\n",
    "    driver = GraphDatabase.driver(uri, auth=(user, password))\n",
    "    return driver\n",
    "\n",
    "# Function to execute a Cypher query\n",
    "def execute_query(driver, query):\n",
    "    with driver.session() as session:\n",
    "        result = session.run(query)\n",
    "        return [record for record in result]\n",
    "\n",
    "# Connect to the Neo4j database\n",
    "uri = \"bolt://localhost:7687\"\n",
    "user = \"neo4j\"\n",
    "password = \"password\"  # Replace 'your_password' with your actual password\n",
    "driver = create_db_connection(uri, user, password)\n",
    "\n",
    "# Strict, Must have Combination Therapies\n",
    "query = \"\"\"MATCH (s:Study)-[:HAS_VARIANT]->(variant),\n",
    "      (s)-[:HAS_THERAPEUTIC]->(therapeutic),\n",
    "      (therapeutic)-[:HAS_COMPONENTS]->(combination)\n",
    "RETURN properties(s) AS Study, \n",
    "       properties(variant) AS VariantProperties,\n",
    "       properties(therapeutic) AS TherapeuticProperties,\n",
    "       COUNT(combination) AS NumberOfComponents,\n",
    "       COLLECT(properties(combination)) AS CombinationTherapyProperties\n",
    "\"\"\"\n",
    "\n",
    "query_therapies = \"\"\"\n",
    "MATCH (s:Study)-[:HAS_VARIANT]->(variant),\n",
    "      (s)-[:HAS_THERAPEUTIC]->(therapeutic)\n",
    "OPTIONAL MATCH (therapeutic)-[:HAS_COMPONENTS]->(combination)\n",
    "RETURN properties(s) AS Study, \n",
    "       properties(variant) AS VariantProperties,\n",
    "       properties(therapeutic) AS TherapeuticProperties,\n",
    "       COUNT(combination) AS NumberOfComponents,\n",
    "       COLLECT(properties(combination)) AS CombinationTherapyProperties\n",
    "\"\"\"\n",
    "\n",
    "# Execute the query\n",
    "result = execute_query(driver, query_therapies)\n",
    "\n",
    "# Close the connection\n",
    "driver.close()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_allele_origin</th>\n",
       "      <th>study_id</th>\n",
       "      <th>study_direction</th>\n",
       "      <th>study_predicate</th>\n",
       "      <th>study_type</th>\n",
       "      <th>variant_mp_score</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>variant_label</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>therapeutic_type</th>\n",
       "      <th>therapeutic_civic_type</th>\n",
       "      <th>therapeutic_id</th>\n",
       "      <th>number_of_components</th>\n",
       "      <th>combination_therapy_components</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>somatic</td>\n",
       "      <td>civic.eid:238</td>\n",
       "      <td>supports</td>\n",
       "      <td>predictsResistanceTo</td>\n",
       "      <td>VariantTherapeuticResponseStudy</td>\n",
       "      <td>406.25</td>\n",
       "      <td>civic.mpid:34</td>\n",
       "      <td>EGFR T790M</td>\n",
       "      <td>[{\"label\": \"missense_variant\", \"system\": \"http...</td>\n",
       "      <td>TherapeuticAgent</td>\n",
       "      <td>None</td>\n",
       "      <td>civic.tid:15</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>somatic</td>\n",
       "      <td>civic.eid:1409</td>\n",
       "      <td>supports</td>\n",
       "      <td>predictsSensitivityTo</td>\n",
       "      <td>VariantTherapeuticResponseStudy</td>\n",
       "      <td>1378.50</td>\n",
       "      <td>civic.mpid:12</td>\n",
       "      <td>BRAF V600E</td>\n",
       "      <td>[{\"label\": \"missense_variant\", \"system\": \"http...</td>\n",
       "      <td>TherapeuticAgent</td>\n",
       "      <td>None</td>\n",
       "      <td>civic.tid:4</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>somatic</td>\n",
       "      <td>civic.eid:1592</td>\n",
       "      <td>supports</td>\n",
       "      <td>predictsSensitivityTo</td>\n",
       "      <td>VariantTherapeuticResponseStudy</td>\n",
       "      <td>406.25</td>\n",
       "      <td>civic.mpid:34</td>\n",
       "      <td>EGFR T790M</td>\n",
       "      <td>[{\"label\": \"missense_variant\", \"system\": \"http...</td>\n",
       "      <td>TherapeuticAgent</td>\n",
       "      <td>None</td>\n",
       "      <td>civic.tid:187</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>somatic</td>\n",
       "      <td>civic.eid:1867</td>\n",
       "      <td>supports</td>\n",
       "      <td>predictsSensitivityTo</td>\n",
       "      <td>VariantTherapeuticResponseStudy</td>\n",
       "      <td>406.25</td>\n",
       "      <td>civic.mpid:34</td>\n",
       "      <td>EGFR T790M</td>\n",
       "      <td>[{\"label\": \"missense_variant\", \"system\": \"http...</td>\n",
       "      <td>TherapeuticAgent</td>\n",
       "      <td>None</td>\n",
       "      <td>civic.tid:187</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>somatic</td>\n",
       "      <td>civic.eid:2994</td>\n",
       "      <td>supports</td>\n",
       "      <td>predictsSensitivityTo</td>\n",
       "      <td>VariantTherapeuticResponseStudy</td>\n",
       "      <td>379.00</td>\n",
       "      <td>civic.mpid:33</td>\n",
       "      <td>EGFR L858R</td>\n",
       "      <td>[{\"label\": \"missense_variant\", \"system\": \"http...</td>\n",
       "      <td>TherapeuticAgent</td>\n",
       "      <td>None</td>\n",
       "      <td>civic.tid:15</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1037</th>\n",
       "      <td>somatic</td>\n",
       "      <td>moa.assertion:961</td>\n",
       "      <td>none</td>\n",
       "      <td>predictsSensitivityTo</td>\n",
       "      <td>VariantTherapeuticResponseStudy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>moa.variant:66</td>\n",
       "      <td>ABL1 p.T315I (Missense)</td>\n",
       "      <td>None</td>\n",
       "      <td>TherapeuticAgent</td>\n",
       "      <td>None</td>\n",
       "      <td>moa.normalize.therapy.rxcui:1364347</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1038</th>\n",
       "      <td>somatic</td>\n",
       "      <td>moa.assertion:963</td>\n",
       "      <td>none</td>\n",
       "      <td>predictsSensitivityTo</td>\n",
       "      <td>VariantTherapeuticResponseStudy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>moa.variant:146</td>\n",
       "      <td>BRAF p.V600K (Missense)</td>\n",
       "      <td>None</td>\n",
       "      <td>TherapeuticAgent</td>\n",
       "      <td>None</td>\n",
       "      <td>moa.normalize.therapy.ncit:C106254</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1039</th>\n",
       "      <td>somatic</td>\n",
       "      <td>moa.assertion:967</td>\n",
       "      <td>none</td>\n",
       "      <td>predictsSensitivityTo</td>\n",
       "      <td>VariantTherapeuticResponseStudy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>moa.variant:254</td>\n",
       "      <td>EGFR p.L858R (Missense)</td>\n",
       "      <td>None</td>\n",
       "      <td>TherapeuticAgent</td>\n",
       "      <td>None</td>\n",
       "      <td>moa.normalize.therapy.rxcui:337525</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1040</th>\n",
       "      <td>somatic</td>\n",
       "      <td>moa.assertion:969</td>\n",
       "      <td>none</td>\n",
       "      <td>predictsSensitivityTo</td>\n",
       "      <td>VariantTherapeuticResponseStudy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>moa.variant:254</td>\n",
       "      <td>EGFR p.L858R (Missense)</td>\n",
       "      <td>None</td>\n",
       "      <td>TherapeuticAgent</td>\n",
       "      <td>None</td>\n",
       "      <td>moa.normalize.therapy.rxcui:328134</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1041</th>\n",
       "      <td>somatic</td>\n",
       "      <td>moa.assertion:973</td>\n",
       "      <td>none</td>\n",
       "      <td>predictsSensitivityTo</td>\n",
       "      <td>VariantTherapeuticResponseStudy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>moa.variant:372</td>\n",
       "      <td>KRAS p.G12C (Missense)</td>\n",
       "      <td>None</td>\n",
       "      <td>CombinationTherapy</td>\n",
       "      <td>None</td>\n",
       "      <td>moa.ctid:5e89NyyfMGR167xjfPQC_QJBxBI9OYZx</td>\n",
       "      <td>2</td>\n",
       "      <td>[{'id': 'moa.normalize.therapy.ncit:C157493', ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1042 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     study_allele_origin           study_id study_direction  \\\n",
       "0                somatic      civic.eid:238        supports   \n",
       "1                somatic     civic.eid:1409        supports   \n",
       "2                somatic     civic.eid:1592        supports   \n",
       "3                somatic     civic.eid:1867        supports   \n",
       "4                somatic     civic.eid:2994        supports   \n",
       "...                  ...                ...             ...   \n",
       "1037             somatic  moa.assertion:961            none   \n",
       "1038             somatic  moa.assertion:963            none   \n",
       "1039             somatic  moa.assertion:967            none   \n",
       "1040             somatic  moa.assertion:969            none   \n",
       "1041             somatic  moa.assertion:973            none   \n",
       "\n",
       "            study_predicate                       study_type  \\\n",
       "0      predictsResistanceTo  VariantTherapeuticResponseStudy   \n",
       "1     predictsSensitivityTo  VariantTherapeuticResponseStudy   \n",
       "2     predictsSensitivityTo  VariantTherapeuticResponseStudy   \n",
       "3     predictsSensitivityTo  VariantTherapeuticResponseStudy   \n",
       "4     predictsSensitivityTo  VariantTherapeuticResponseStudy   \n",
       "...                     ...                              ...   \n",
       "1037  predictsSensitivityTo  VariantTherapeuticResponseStudy   \n",
       "1038  predictsSensitivityTo  VariantTherapeuticResponseStudy   \n",
       "1039  predictsSensitivityTo  VariantTherapeuticResponseStudy   \n",
       "1040  predictsSensitivityTo  VariantTherapeuticResponseStudy   \n",
       "1041  predictsSensitivityTo  VariantTherapeuticResponseStudy   \n",
       "\n",
       "      variant_mp_score       variant_id            variant_label  \\\n",
       "0               406.25    civic.mpid:34               EGFR T790M   \n",
       "1              1378.50    civic.mpid:12               BRAF V600E   \n",
       "2               406.25    civic.mpid:34               EGFR T790M   \n",
       "3               406.25    civic.mpid:34               EGFR T790M   \n",
       "4               379.00    civic.mpid:33               EGFR L858R   \n",
       "...                ...              ...                      ...   \n",
       "1037               NaN   moa.variant:66  ABL1 p.T315I (Missense)   \n",
       "1038               NaN  moa.variant:146  BRAF p.V600K (Missense)   \n",
       "1039               NaN  moa.variant:254  EGFR p.L858R (Missense)   \n",
       "1040               NaN  moa.variant:254  EGFR p.L858R (Missense)   \n",
       "1041               NaN  moa.variant:372   KRAS p.G12C (Missense)   \n",
       "\n",
       "                                           variant_type    therapeutic_type  \\\n",
       "0     [{\"label\": \"missense_variant\", \"system\": \"http...    TherapeuticAgent   \n",
       "1     [{\"label\": \"missense_variant\", \"system\": \"http...    TherapeuticAgent   \n",
       "2     [{\"label\": \"missense_variant\", \"system\": \"http...    TherapeuticAgent   \n",
       "3     [{\"label\": \"missense_variant\", \"system\": \"http...    TherapeuticAgent   \n",
       "4     [{\"label\": \"missense_variant\", \"system\": \"http...    TherapeuticAgent   \n",
       "...                                                 ...                 ...   \n",
       "1037                                               None    TherapeuticAgent   \n",
       "1038                                               None    TherapeuticAgent   \n",
       "1039                                               None    TherapeuticAgent   \n",
       "1040                                               None    TherapeuticAgent   \n",
       "1041                                               None  CombinationTherapy   \n",
       "\n",
       "     therapeutic_civic_type                             therapeutic_id  \\\n",
       "0                      None                               civic.tid:15   \n",
       "1                      None                                civic.tid:4   \n",
       "2                      None                              civic.tid:187   \n",
       "3                      None                              civic.tid:187   \n",
       "4                      None                               civic.tid:15   \n",
       "...                     ...                                        ...   \n",
       "1037                   None        moa.normalize.therapy.rxcui:1364347   \n",
       "1038                   None         moa.normalize.therapy.ncit:C106254   \n",
       "1039                   None         moa.normalize.therapy.rxcui:337525   \n",
       "1040                   None         moa.normalize.therapy.rxcui:328134   \n",
       "1041                   None  moa.ctid:5e89NyyfMGR167xjfPQC_QJBxBI9OYZx   \n",
       "\n",
       "      number_of_components                     combination_therapy_components  \n",
       "0                        0                                                 []  \n",
       "1                        0                                                 []  \n",
       "2                        0                                                 []  \n",
       "3                        0                                                 []  \n",
       "4                        0                                                 []  \n",
       "...                    ...                                                ...  \n",
       "1037                     0                                                 []  \n",
       "1038                     0                                                 []  \n",
       "1039                     0                                                 []  \n",
       "1040                     0                                                 []  \n",
       "1041                     2  [{'id': 'moa.normalize.therapy.ncit:C157493', ...  \n",
       "\n",
       "[1042 rows x 14 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "data = []\n",
    "for record in result:\n",
    "    row = {\n",
    "        'study_allele_origin': record['Study'].get('alleleOrigin', None),\n",
    "        'study_id': record['Study']['id'],\n",
    "        'study_direction': record['Study']['direction'],\n",
    "        'study_predicate': record['Study']['predicate'],\n",
    "        'study_type': record['Study']['type'],\n",
    "        'variant_mp_score': record['VariantProperties'].get('civic_molecular_profile_score', None),\n",
    "        'variant_id': record['VariantProperties']['id'],\n",
    "        'variant_label': record['VariantProperties']['label'],\n",
    "        'variant_type': record['VariantProperties'].get('variant_types',None),\n",
    "        'therapeutic_type': record['TherapeuticProperties']['type'],\n",
    "        'therapeutic_civic_type': record['TherapeuticProperties'].get('civic_therapy_interaction_type', None),\n",
    "        'therapeutic_id': record['TherapeuticProperties']['id'],\n",
    "        'number_of_components': record.get('NumberOfComponents', None),\n",
    "        'combination_therapy_components': record.get('CombinationTherapyProperties', None)\n",
    "    }\n",
    "\n",
    "    data.append(row)\n",
    "\n",
    "df = pd.DataFrame(data)\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inspect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count     815.000000\n",
       "mean      176.373620\n",
       "std       382.819215\n",
       "min         0.500000\n",
       "25%         7.500000\n",
       "50%        20.000000\n",
       "75%        96.500000\n",
       "max      1378.500000\n",
       "Name: variant_mp_score, dtype: float64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['variant_mp_score'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "variant_label\n",
       "BRAF V600E                 70\n",
       "EGFR L858R                 36\n",
       "BRAF p.V600E (Missense)    31\n",
       "EGFR T790M                 27\n",
       "PIK3CA H1047R              22\n",
       "                           ..\n",
       "KDR R961W                   1\n",
       "ERBB2 G778_S779insLPS       1\n",
       "ERBB2 G776delinsLC          1\n",
       "ERBB2 G776delinsCV          1\n",
       "IDH1 p.R132L (Missense)     1\n",
       "Name: count, Length: 474, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['variant_label'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "number_of_components\n",
       "0    922\n",
       "2    101\n",
       "3     19\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['number_of_components'].value_counts()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
